+++
# About/Biography widget.
widget = "about"
active = true
date = 2016-04-20T00:00:00

# Order that this section will appear in.
weight = 5

# List your academic interests.
[interests]
  interests = [
    "(Epi)genomics",
    "Education",
    "Machine learning",
    "Unix",
    "rstats"
  ]

# List your qualifications (such as academic degrees).
[[education.courses]]
  course = "PhD in Genetics and Genomics"
  institution = "University of Florida"
  year = 2014

[[education.courses]]
  course = "BS in Biotechnology"
  institution = "Shanghai Jiaotong University"
  year = 2008

 
+++

# About me

I am a computational biologist working on genomics, epigenomics and transcriptomics. I use R primary for data wrangling and visualization in the [tidyverse](https://www.tidyverse.org/) ecosystem; I use python for writing [Snakemake](https://snakemake.readthedocs.io/en/stable/) workflows; I am a unix geek learning shell tricks almost every month; I care about reproducible research and open science.

Being trained in a wet lab in the University of Florida during my PhD in [Dr.Jianrong Lu's lab](https://biochem.med.ufl.edu/research/primary-faculty/jianrong-lu/) has established my solid knowledge and skills in experimental molecular cancer biology. Self-teaching and postdoctoral training in [Dr.Roel Verhaak's lab](https://www.jax.org/research-and-faculty/faculty/roel-verhaak) has extended my bioinformatics skills in integrating analysis of TB size sequencing data sets. Verhaak lab is well known for studying genomic alterations of brain tumor by analyzing large panels of RNA-seq and DNA-seq data. I gained extensive experience in handling large-scale genomic data and pipelining workflows. I also gained intimate familiarities with public data sets such as [ENCODE](https://www.encodeproject.org/), [TCGA](https://portal.gdc.cancer.gov/) and [CCLE](http://www.broadinstitute.org/ccle/home). I have put my analysis notes and snakemake pipelines for processing whole-exome, whole-genome DNAseq, RNAseq, single-cell RNAseq, ChIP-seq, ATACseq and RRBS data in my [github repos](https://github.com/crazyhottommy). 

Roel moved to Jackson lab for Genomics Medicine in October 2016. To apply my skills to translational lung cancer research, I joined [Dr. Andrew Futreal](https://gsbs.uth.edu/faculty/faculty-directory/faculty-profiles.htm?id=bc44b01e-38b8-4e9f-ab8b-627c05c4b708) and [Dr.Jianjun Zhangâ€™s](http://faculty.mdanderson.org/Jianjun_Zhang/) lab shortly to study tumor heterogeneity in lung cancer by analyzing multi-region whole-exome sequencing data and DNA methylation array data. I have been working as a research scientist in MD Anderson at [Kunal Rai's lab](http://railab.org/people.html) studying cancer epigenomics.

I joined [Harvard Faculty of Arts and Sciences informatics](https://informatics.fas.harvard.edu) team on October 1, 2018 as a bioinformatics scientist working closely with [Dr.Dulac lab](https://www.dulaclab.com)s to catalog and understand the diversity and function of cell types in the mouse brain using single-cell RNA-seq and other cutting edge techniques.

I also have a great interest in promoting open science and reforming bioinformatics education. I frequently share my thoughts on [twitter](https://twitter.com/tangming2005) and tips in my [blog post](http://crazyhottommy.blogspot.com/). I am a certified instructor for the [carpentries](https://carpentries.org/).


