<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.37" />
  <meta name="author" content="Ming Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="Using nested dataframe and list column has transformed my way of data wrangling in R. For more on this topic, I highly recommend purrr tutorial from Jenney Bryan.
In this post, I am going to show you how I use this to solve a problem for adding pct_in column from the differential scRNAseq result table.
I am going to use presto for differential gene expression test. presto performs a fast Wilcoxon rank sum test and auROC analysis.">

  
  <link rel="alternate" hreflang="en-us" href="/post/add-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/add-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="DNA confesses Data speak">
  <meta property="og:url" content="/post/add-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column/">
  <meta property="og:title" content="add pct_in for each cluster for scRNAseq result table using list column | DNA confesses Data speak">
  <meta property="og:description" content="Using nested dataframe and list column has transformed my way of data wrangling in R. For more on this topic, I highly recommend purrr tutorial from Jenney Bryan.
In this post, I am going to show you how I use this to solve a problem for adding pct_in column from the differential scRNAseq result table.
I am going to use presto for differential gene expression test. presto performs a fast Wilcoxon rank sum test and auROC analysis.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2020-02-04T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2020-02-04T00:00:00&#43;00:00">
  

  
  

  <title>add pct_in for each cluster for scRNAseq result table using list column | DNA confesses Data speak</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">DNA confesses Data speak</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>CV</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">add pct_in for each cluster for scRNAseq result table using list column</h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2020-02-04 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Feb 4, 2020
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    9 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/add-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/bioinformatics/">bioinformatics</a
    >, 
    
    <a href="/categories/scrnaseq/">scRNAseq</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=add%20pct_in%20for%20each%20cluster%20for%20scRNAseq%20result%20table%20using%20list%20column&amp;url=%2fpost%2fadd-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fadd-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fadd-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column%2f&amp;title=add%20pct_in%20for%20each%20cluster%20for%20scRNAseq%20result%20table%20using%20list%20column"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fadd-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column%2f&amp;title=add%20pct_in%20for%20each%20cluster%20for%20scRNAseq%20result%20table%20using%20list%20column"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=add%20pct_in%20for%20each%20cluster%20for%20scRNAseq%20result%20table%20using%20list%20column&amp;body=%2fpost%2fadd-pct-in-for-each-cluster-for-scrnaseq-result-table-using-list-column%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <p>Using nested dataframe and list column has transformed my way of data wrangling in R. For more on this topic, I highly recommend <a href="https://jennybc.github.io/purrr-tutorial/index.html">purrr tutorial</a> from Jenney Bryan.</p>
<p>In this post, I am going to show you how I use this to solve a problem for adding <code>pct_in</code> column from the differential scRNAseq result table.</p>
<p>I am going to use <a href="https://github.com/immunogenomics/presto"><code>presto</code></a> for differential gene expression test. <code>presto</code> performs a fast Wilcoxon rank sum test and auROC analysis. It can be used for differential accessible region test for scATACseq data as well. Because scATACseq data can have over 800k regions in my hand, I found it is much faster than <code>Seurat</code> and also gives sensible results. Using presto also gives you all the genes/regions without filtering. This is particularly useful if you want to run GSEA which requires all genes as input. see <a href="https://crazyhottommy.github.io/scRNA-seq-workshop-Fall-2019/scRNAseq_workshop_3.html">this part</a> for our scRNAseq workshop.</p>
<p>Let’s download some scRNAseq example data from <a href="https://satijalab.org/seurat/v3.1/atacseq_integration_vignette.html" class="uri">https://satijalab.org/seurat/v3.1/atacseq_integration_vignette.html</a>.</p>
<p><code>curl -L https://www.dropbox.com/s/3f3p5nxrn5b3y4y/pbmc_10k_v3.rds\?dl\=1 -o pbmc_10k_v3.rds</code></p>
<div id="read-into-r" class="section level3">
<h3>read into R</h3>
<pre class="r"><code># install_github(&#39;immunogenomics/presto&#39;)
library(presto)
library(Seurat)
library(dplyr)
library(tibble)
library(furrr)
library(tictoc)
pbmc&lt;- readRDS(&quot;~/pbmc_10k_v3.rds&quot;)

head(wilcoxauc(pbmc, &quot;RNA_snn_res.0.4&quot; ))</code></pre>
<pre><code>##      feature group     avgExpr         logFC statistic       auc
## 1 AL627309.1     0 0.004455649 -0.0001475212   9648856 0.5007586
## 2 AL669831.5     0 0.058653109  0.0155762453   9974323 0.5176497
## 3     FAM87B     0 0.001185081  0.0007726323   9649065 0.5007694
## 4  LINC00115     0 0.023454801 -0.0001696563   9701960 0.5035145
## 5     FAM41C     0 0.025523520  0.0035611306   9745564 0.5057775
## 6 AL645608.3     0 0.000766896  0.0003039512   9642626 0.5004352
##           pval         padj    pct_in    pct_out
## 1 3.450509e-01 4.033191e-01 0.6350267 0.48136646
## 2 7.957294e-12 2.279022e-11 8.3221925 4.58074534
## 3 2.429577e-02 3.743801e-02 0.2005348 0.04658385
## 4 5.546520e-02 8.014953e-02 3.3422460 2.59316770
## 5 1.339544e-03 2.430893e-03 3.5427807 2.34472050
## 6 1.485920e-01 1.932465e-01 0.1336898 0.04658385</code></pre>
<p>by default, <code>presto</code> and <code>Seurat</code> compare a gene in cells of one group versus all other groups of cells and calculate the
statistics. In the output, you see <code>pct_in</code> and <code>pct_out</code> columns which show the percentage of cells express this gene in the <code>in</code> group and perecentage of cells express this gene in the <code>out</code> groups. What if you want to know <code>pct_out</code> in each of the group? How do you add that information to the dataframe? In addtion, you may also want to add the number of cells in each cluster into the dataframe.</p>
<pre class="r"><code>res&lt;- wilcoxauc(pbmc, &quot;RNA_snn_res.0.4&quot; )

## how many genes in the result?
length(unique(res$feature))</code></pre>
<pre><code>## [1] 19089</code></pre>
<pre class="r"><code>## for each group we have the same number of genes
count(res, group) %&gt;% arrange(as.numeric(group))</code></pre>
<pre><code>## # A tibble: 13 x 2
##    group     n
##    &lt;chr&gt; &lt;int&gt;
##  1 0     19089
##  2 1     19089
##  3 2     19089
##  4 3     19089
##  5 4     19089
##  6 5     19089
##  7 6     19089
##  8 7     19089
##  9 8     19089
## 10 9     19089
## 11 10    19089
## 12 11    19089
## 13 12    19089</code></pre>
<p>get a dataframe for number of cells in each group</p>
<pre class="r"><code>(cell_number&lt;- pbmc@meta.data %&gt;%
  count(RNA_snn_res.0.4) %&gt;%
  dplyr::rename(group = RNA_snn_res.0.4, cell_number = n))</code></pre>
<pre><code>## # A tibble: 13 x 2
##    group cell_number
##    &lt;fct&gt;       &lt;int&gt;
##  1 0            2992
##  2 1            1596
##  3 2            1047
##  4 3             959
##  5 4             592
##  6 5             544
##  7 6             460
##  8 7             383
##  9 8             337
## 10 9             328
## 11 10             74
## 12 11             68
## 13 12             52</code></pre>
<p>Let’s nest the dataframe by gene</p>
<pre class="r"><code>res_nest&lt;- res %&gt;%
  group_by(feature) %&gt;% 
  tidyr::nest()

res_nest</code></pre>
<pre><code>## # A tibble: 19,089 x 2
## # Groups:   feature [19,089]
##    feature    data                                                         
##    &lt;chr&gt;      &lt;S3: vctrs_list_of&gt;                                          
##  1 AL627309.1 0                    , 1                    , 10            …
##  2 AL669831.5 0                   , 1                   , 10              …
##  3 FAM87B     0                    , 1                    , 10            …
##  4 LINC00115  0                    , 1                    , 10            …
##  5 FAM41C     0                   , 1                   , 10              …
##  6 AL645608.3 0                    , 1                    , 10            …
##  7 SAMD11     0                    , 1                    , 10            …
##  8 NOC2L      0                   , 1                   , 10              …
##  9 KLHL17     0                   , 1                   , 10              …
## 10 PLEKHN1    0                    , 1                    , 10            …
## # … with 19,079 more rows</code></pre>
<p><code>res_nest</code> is a nested dataframe with a list column named <code>data</code>. Let’s check the first entry of this list.</p>
<pre class="r"><code>df&lt;- res_nest$data[[1]] %&gt;% arrange(as.numeric(group))
head(df)  </code></pre>
<pre><code>## # A tibble: 6 x 9
##   group avgExpr     logFC statistic   auc   pval  padj pct_in pct_out
##   &lt;chr&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
## 1 0     0.00446 -0.000148  9648856. 0.501 0.345  0.403  0.635   0.481
## 2 1     0.00206 -0.00300   6232089  0.498 0.0916 0.162  0.251   0.587
## 3 2     0.00285 -0.00192   4377551  0.499 0.251  0.364  0.287   0.561
## 4 3     0.00685  0.00255   4067216. 0.501 0.661  0.723  0.626   0.519
## 5 4     0.00650  0.00208   2620733  0.501 0.612  0.719  0.676   0.520
## 6 5     0.00834  0.00401   2422859  0.501 0.492  0.648  0.735   0.518</code></pre>
<p>Now, we can collect the <code>pct_in</code> for this gene from <code>df</code>.</p>
<pre class="r"><code>(df&lt;- df %&gt;% 
  left_join(cell_number, by = c(&quot;group&quot; = &quot;group&quot;)) %&gt;%
  mutate(pct_in_group = paste(group, pct_in, sep= &quot;_&quot;)))</code></pre>
<pre><code>## Warning: Column `group` joining character vector and factor, coercing into
## character vector</code></pre>
<pre><code>## # A tibble: 13 x 11
##    group avgExpr    logFC statistic   auc   pval  padj pct_in pct_out
##    &lt;chr&gt;   &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
##  1 0     0.00446 -1.48e-4  9648856. 0.501 0.345  0.403  0.635   0.481
##  2 1     0.00206 -3.00e-3  6232089  0.498 0.0916 0.162  0.251   0.587
##  3 2     0.00285 -1.92e-3  4377551  0.499 0.251  0.364  0.287   0.561
##  4 3     0.00685  2.55e-3  4067216. 0.501 0.661  0.723  0.626   0.519
##  5 4     0.00650  2.08e-3  2620733  0.501 0.612  0.719  0.676   0.520
##  6 5     0.00834  4.01e-3  2422859  0.501 0.492  0.648  0.735   0.518
##  7 6     0.00816  3.78e-3  2070947  0.502 0.303  0.507  0.870   0.513
##  8 7     0.00245 -2.20e-3  1728031  0.499 0.460  0.685  0.261   0.541
##  9 8     0       -4.73e-3  1524082. 0.497 0.172  0.348  0       0.550
## 10 9     0.00533  7.97e-4  1498956. 0.502 0.333  0.524  0.915   0.516
## 11 10    0.00609  1.55e-3   349088. 0.504 0.333  0.586  1.35    0.524
## 12 11    0       -4.59e-3   316676  0.497 0.546  0.848  0       0.534
## 13 12    0.0295   2.50e-2   247320. 0.507 0.163  0.437  1.92    0.522
## # … with 2 more variables: cell_number &lt;int&gt;, pct_in_group &lt;chr&gt;</code></pre>
<pre class="r"><code># interleave the pct_in and number_in 
pct_in_groups&lt;- df$pct_in
num_in_groups&lt;- df$cell_number
names_pct_in_groups&lt;-  paste(df$group,&quot;pct_in&quot;, sep = &quot;_&quot;)
names_num_in_groups&lt;- paste(df$group, &quot;cell_num&quot;, sep= &quot;_&quot;)
# https://stackoverflow.com/questions/16443260/interleave-lists-in-r
out&lt;- c(rbind(num_in_groups, pct_in_groups))
names(out)&lt;- c(rbind(names_num_in_groups, names_pct_in_groups))
out&lt;- bind_rows(out)
out</code></pre>
<pre><code>## # A tibble: 1 x 26
##   `0_cell_num` `0_pct_in` `1_cell_num` `1_pct_in` `2_cell_num` `2_pct_in`
##          &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;
## 1         2992      0.635         1596      0.251         1047      0.287
## # … with 20 more variables: `3_cell_num` &lt;dbl&gt;, `3_pct_in` &lt;dbl&gt;,
## #   `4_cell_num` &lt;dbl&gt;, `4_pct_in` &lt;dbl&gt;, `5_cell_num` &lt;dbl&gt;,
## #   `5_pct_in` &lt;dbl&gt;, `6_cell_num` &lt;dbl&gt;, `6_pct_in` &lt;dbl&gt;,
## #   `7_cell_num` &lt;dbl&gt;, `7_pct_in` &lt;dbl&gt;, `8_cell_num` &lt;dbl&gt;,
## #   `8_pct_in` &lt;dbl&gt;, `9_cell_num` &lt;dbl&gt;, `9_pct_in` &lt;dbl&gt;,
## #   `10_cell_num` &lt;dbl&gt;, `10_pct_in` &lt;dbl&gt;, `11_cell_num` &lt;dbl&gt;,
## #   `11_pct_in` &lt;dbl&gt;, `12_cell_num` &lt;dbl&gt;, `12_pct_in` &lt;dbl&gt;</code></pre>
<p>We have 13 groups of cells, so we get a tibble of 1 x 26 with number of cells and percentage of cells for each group in each column. We now only need to <code>cbind</code> this info back to each gene.</p>
<p>Let’s write a function</p>
<pre class="r"><code>add_pct_in&lt;- function(df, cell_number){
  df&lt;- df %&gt;% 
  left_join(cell_number, by = c(&quot;group&quot; = &quot;group&quot;)) %&gt;%
  mutate(pct_in_group = paste(group, pct_in, sep= &quot;_&quot;))
  
  pct_in_groups&lt;- df$pct_in
  num_in_groups&lt;- df$cell_number
  names_pct_in_groups&lt;-  paste(df$group,&quot;pct_in&quot;, sep = &quot;_&quot;)
  names_num_in_groups&lt;- paste(df$group, &quot;cell_num&quot;, sep= &quot;_&quot;)
  # https://stackoverflow.com/questions/16443260/interleave-lists-in-r
  out&lt;- c(rbind(num_in_groups, pct_in_groups))
  names(out)&lt;- c(rbind(names_num_in_groups, names_pct_in_groups))
  out&lt;- bind_rows(out)
  return(out)
}

## test this function for one gene
add_pct_in(df = res_nest$data[[1]], cell_number = cell_number )</code></pre>
<pre><code>## Warning: Column `group` joining character vector and factor, coercing into
## character vector</code></pre>
<pre><code>## # A tibble: 1 x 26
##   `0_cell_num` `0_pct_in` `1_cell_num` `1_pct_in` `10_cell_num` `10_pct_in`
##          &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;         &lt;dbl&gt;       &lt;dbl&gt;
## 1         2992      0.635         1596      0.251            74        1.35
## # … with 20 more variables: `11_cell_num` &lt;dbl&gt;, `11_pct_in` &lt;dbl&gt;,
## #   `12_cell_num` &lt;dbl&gt;, `12_pct_in` &lt;dbl&gt;, `2_cell_num` &lt;dbl&gt;,
## #   `2_pct_in` &lt;dbl&gt;, `3_cell_num` &lt;dbl&gt;, `3_pct_in` &lt;dbl&gt;,
## #   `4_cell_num` &lt;dbl&gt;, `4_pct_in` &lt;dbl&gt;, `5_cell_num` &lt;dbl&gt;,
## #   `5_pct_in` &lt;dbl&gt;, `6_cell_num` &lt;dbl&gt;, `6_pct_in` &lt;dbl&gt;,
## #   `7_cell_num` &lt;dbl&gt;, `7_pct_in` &lt;dbl&gt;, `8_cell_num` &lt;dbl&gt;,
## #   `8_pct_in` &lt;dbl&gt;, `9_cell_num` &lt;dbl&gt;, `9_pct_in` &lt;dbl&gt;</code></pre>
<p>Because we are going to apply this function to over 10,000 genes, I am going to use the parallized <code>purrr</code>: <code>furrr</code>.
<a href="https://github.com/DavisVaughan/furrr" class="uri">https://github.com/DavisVaughan/furrr</a></p>
<pre class="r"><code>plan(multiprocess, workers = 8)

# this will start 8 workers, but each worker will consume 20Mb memory 
print(object.size(res), units= &quot;Mb&quot;)</code></pre>
<pre><code>## 20 Mb</code></pre>
<pre class="r"><code>tic()
info&lt;- furrr::future_map_dfr(res_nest$data, ~ add_pct_in(df= .x, cell_number = cell_number)) 
toc()</code></pre>
<pre><code>## 9.007 sec elapsed</code></pre>
<pre class="r"><code>head(info)</code></pre>
<pre><code>## # A tibble: 6 x 26
##   `0_cell_num` `0_pct_in` `1_cell_num` `1_pct_in` `10_cell_num` `10_pct_in`
##          &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;         &lt;dbl&gt;       &lt;dbl&gt;
## 1         2992      0.635         1596     0.251             74        1.35
## 2         2992      8.32          1596     3.57              74       20.3 
## 3         2992      0.201         1596     0.0627            74        0   
## 4         2992      3.34          1596     2.26              74        9.46
## 5         2992      3.54          1596     1.63              74        9.46
## 6         2992      0.134         1596     0                 74        0   
## # … with 20 more variables: `11_cell_num` &lt;dbl&gt;, `11_pct_in` &lt;dbl&gt;,
## #   `12_cell_num` &lt;dbl&gt;, `12_pct_in` &lt;dbl&gt;, `2_cell_num` &lt;dbl&gt;,
## #   `2_pct_in` &lt;dbl&gt;, `3_cell_num` &lt;dbl&gt;, `3_pct_in` &lt;dbl&gt;,
## #   `4_cell_num` &lt;dbl&gt;, `4_pct_in` &lt;dbl&gt;, `5_cell_num` &lt;dbl&gt;,
## #   `5_pct_in` &lt;dbl&gt;, `6_cell_num` &lt;dbl&gt;, `6_pct_in` &lt;dbl&gt;,
## #   `7_cell_num` &lt;dbl&gt;, `7_pct_in` &lt;dbl&gt;, `8_cell_num` &lt;dbl&gt;,
## #   `8_pct_in` &lt;dbl&gt;, `9_cell_num` &lt;dbl&gt;, `9_pct_in` &lt;dbl&gt;</code></pre>
<pre class="r"><code>## cbind back to the nested dataframe
bind_cols(res_nest, info) %&gt;%
  head()</code></pre>
<pre><code>## # A tibble: 6 x 28
## # Groups:   feature [6]
##   feature data  `0_cell_num` `0_pct_in` `1_cell_num` `1_pct_in`
##   &lt;chr&gt;   &lt;S3:&gt;        &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;
## 1 AL6273… 0   …         2992      0.635         1596     0.251 
## 2 AL6698… 0   …         2992      8.32          1596     3.57  
## 3 FAM87B  0   …         2992      0.201         1596     0.0627
## 4 LINC00… 0   …         2992      3.34          1596     2.26  
## 5 FAM41C  0   …         2992      3.54          1596     1.63  
## 6 AL6456… 0   …         2992      0.134         1596     0     
## # … with 22 more variables: `10_cell_num` &lt;dbl&gt;, `10_pct_in` &lt;dbl&gt;,
## #   `11_cell_num` &lt;dbl&gt;, `11_pct_in` &lt;dbl&gt;, `12_cell_num` &lt;dbl&gt;,
## #   `12_pct_in` &lt;dbl&gt;, `2_cell_num` &lt;dbl&gt;, `2_pct_in` &lt;dbl&gt;,
## #   `3_cell_num` &lt;dbl&gt;, `3_pct_in` &lt;dbl&gt;, `4_cell_num` &lt;dbl&gt;,
## #   `4_pct_in` &lt;dbl&gt;, `5_cell_num` &lt;dbl&gt;, `5_pct_in` &lt;dbl&gt;,
## #   `6_cell_num` &lt;dbl&gt;, `6_pct_in` &lt;dbl&gt;, `7_cell_num` &lt;dbl&gt;,
## #   `7_pct_in` &lt;dbl&gt;, `8_cell_num` &lt;dbl&gt;, `8_pct_in` &lt;dbl&gt;,
## #   `9_cell_num` &lt;dbl&gt;, `9_pct_in` &lt;dbl&gt;</code></pre>
<p>Finally, we can unnest the dataframe</p>
<pre class="r"><code>bind_cols(res_nest, info) %&gt;% 
  tidyr::unnest() %&gt;%
  head()</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(data)`</code></pre>
<pre><code>## # A tibble: 6 x 36
## # Groups:   feature [1]
##   feature group avgExpr    logFC statistic   auc   pval  padj pct_in
##   &lt;chr&gt;   &lt;chr&gt;   &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
## 1 AL6273… 0     0.00446 -1.48e-4  9648856. 0.501 0.345  0.403  0.635
## 2 AL6273… 1     0.00206 -3.00e-3  6232089  0.498 0.0916 0.162  0.251
## 3 AL6273… 10    0.00609  1.55e-3   349088. 0.504 0.333  0.586  1.35 
## 4 AL6273… 11    0       -4.59e-3   316676  0.497 0.546  0.848  0    
## 5 AL6273… 12    0.0295   2.50e-2   247320. 0.507 0.163  0.437  1.92 
## 6 AL6273… 2     0.00285 -1.92e-3  4377551  0.499 0.251  0.364  0.287
## # … with 27 more variables: pct_out &lt;dbl&gt;, `0_cell_num` &lt;dbl&gt;,
## #   `0_pct_in` &lt;dbl&gt;, `1_cell_num` &lt;dbl&gt;, `1_pct_in` &lt;dbl&gt;,
## #   `10_cell_num` &lt;dbl&gt;, `10_pct_in` &lt;dbl&gt;, `11_cell_num` &lt;dbl&gt;,
## #   `11_pct_in` &lt;dbl&gt;, `12_cell_num` &lt;dbl&gt;, `12_pct_in` &lt;dbl&gt;,
## #   `2_cell_num` &lt;dbl&gt;, `2_pct_in` &lt;dbl&gt;, `3_cell_num` &lt;dbl&gt;,
## #   `3_pct_in` &lt;dbl&gt;, `4_cell_num` &lt;dbl&gt;, `4_pct_in` &lt;dbl&gt;,
## #   `5_cell_num` &lt;dbl&gt;, `5_pct_in` &lt;dbl&gt;, `6_cell_num` &lt;dbl&gt;,
## #   `6_pct_in` &lt;dbl&gt;, `7_cell_num` &lt;dbl&gt;, `7_pct_in` &lt;dbl&gt;,
## #   `8_cell_num` &lt;dbl&gt;, `8_pct_in` &lt;dbl&gt;, `9_cell_num` &lt;dbl&gt;,
## #   `9_pct_in` &lt;dbl&gt;</code></pre>
<p>There are possiblely other easier ways to achieve the same result. Please share your thoughts in the comments!</p>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/scrnaseq/">scRNAseq</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/mixing-mouse-and-human-10x-single-cell-rnaseq-data/">Mixing mouse and human 10x single cell RNAseq data</a></li>
        
        <li><a href="/post/negative-bionomial-distribution-in-single-cell-rnaseq/">negative bionomial distribution in (single-cell) RNAseq </a></li>
        
        <li><a href="/post/run-rstudio-server-with-singularity-on-hpc/">Run Rstudio server with singularity on HPC</a></li>
        
        <li><a href="/post/understand-10x-scrnaseq-and-scatac-fastqs/">understand 10x scRNAseq and scATAC fastqs</a></li>
        
        <li><a href="/project/evaluating-scrnaseq-cluster/">Evaluating single cell RNAseq cluster stability</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/the-end-of-2019/" rel="prev">The end of 2019</a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

