<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tommy Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="To not miss a post like this, sign up for my newsletter to learn computational biology and bioinformatics.
What is partial correlation Partial correlation measures the relationship between two variables while controlling for the effect of one or more other variables.
Suppose you want to know how X and Y are related, independent of how both are influenced by Z. Partial correlation helps answer:
 If we remove the influence of Z, is there still a connection between X and Y?">

  
  <link rel="alternate" hreflang="en-us" href="/post/partial-cor/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="Chatomics">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="Chatomics">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/partial-cor/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="Chatomics">
  <meta property="og:url" content="/post/partial-cor/">
  <meta property="og:title" content="How to calculate partial correlation controlling cancer types | Chatomics">
  <meta property="og:description" content="To not miss a post like this, sign up for my newsletter to learn computational biology and bioinformatics.
What is partial correlation Partial correlation measures the relationship between two variables while controlling for the effect of one or more other variables.
Suppose you want to know how X and Y are related, independent of how both are influenced by Z. Partial correlation helps answer:
 If we remove the influence of Z, is there still a connection between X and Y?">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2025-05-22T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2025-05-22T00:00:00&#43;00:00">
  

  
  

  <title>How to calculate partial correlation controlling cancer types | Chatomics</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">Chatomics</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>Newsletter</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">How to calculate partial correlation controlling cancer types</h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2025-05-22 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      May 22, 2025
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Ming Tommy Tang">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    14 min read
  </span>
  

  
  
  <span class="middot-divider"></span>
  <a href="/post/partial-cor/#disqus_thread"></a>
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/tutorial/">tutorial</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=How%20to%20calculate%20partial%20correlation%20controlling%20cancer%20types&amp;url=%2fpost%2fpartial-cor%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fpartial-cor%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fpartial-cor%2f&amp;title=How%20to%20calculate%20partial%20correlation%20controlling%20cancer%20types"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fpartial-cor%2f&amp;title=How%20to%20calculate%20partial%20correlation%20controlling%20cancer%20types"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=How%20to%20calculate%20partial%20correlation%20controlling%20cancer%20types&amp;body=%2fpost%2fpartial-cor%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <p><strong>To not miss a post like this, sign up for my <a href="https://divingintogeneticsandgenomics.ck.page/profile">newsletter</a> to learn computational
biology and bioinformatics.</strong></p>
<div id="what-is-partial-correlation" class="section level3">
<h3>What is partial correlation</h3>
<p>Partial correlation measures the relationship between two variables while controlling for the effect of one or more other variables.</p>
<p>Suppose you want to know how X and Y are related, independent of how both are influenced by Z. Partial correlation helps answer:</p>
<blockquote>
<p>If we remove the influence of Z, is there still a connection between X and Y?</p>
</blockquote>
<p>What does it have to do with Bioinformatics?</p>
<p>You are studying the relationship between two genes:</p>
<p>Gene A and Gene B
You observe a high correlation between their expression levels across many samples.</p>
<p>But… both genes might be regulated by Transcription Factor X.</p>
<p>So, is the correlation between Gene A and Gene B direct, or is it just because both are influenced by TF X?</p>
<p>Use partial correlation to test the relationship between Gene A and Gene B, controlling for TF X.</p>
</div>
<div id="a-dummy-example-in-calculating-pearson-correlation" class="section level2">
<h2>A dummy example in calculating pearson correlation</h2>
<p>It is easy to calculate correlation in R with the <code>cor</code> function.</p>
<p>To calculate partial correlation, we will turn to linear regression. What’s the relationship
of linear regression with correlation?</p>
<p>I encourage everyone read this <a href="https://lindeloev.github.io/tests-as-linear/">Common statistical tests are linear models</a>.</p>
<blockquote>
<p>It shows the linear models underlying common parametric and “non-parametric” tests.</p>
</blockquote>
<p>I will walk you through an example in the link above:</p>
<pre class="r"><code># Load packages for data handling and plotting
library(tidyverse)
library(patchwork)
library(broom) # tidy model output
library(kableExtra) # nice table

# Reproducible &quot;random&quot; results
set.seed(123)

# Generate normal data with known parameters
rnorm_fixed&lt;- function(N, mu = 0, sd = 1) scale(rnorm(N)) * sd + mu

y&lt;- c(rnorm(15), exp(rnorm(15)), runif(20, min = -3, max = 0))  # Almost zero mean, not normal

x&lt;- rnorm_fixed(50, mu = 0, sd = 1)  # Used in correlation where this is on x-axis

y2&lt;- x * 2 + rnorm(50)

# Long format data with indicator
value = c(y, y2)
group = rep(c(&#39;y1&#39;, &#39;y2&#39;), each = 50)

value</code></pre>
<pre><code>#&gt;   [1] -0.56047565 -0.23017749  1.55870831  0.07050839  0.12928774  1.71506499
#&gt;   [7]  0.46091621 -1.26506123 -0.68685285 -0.44566197  1.22408180  0.35981383
#&gt;  [13]  0.40077145  0.11068272 -0.55584113  5.97099235  1.64518111  0.13992942
#&gt;  [19]  2.01648501  0.62326006  0.34375582  0.80414561  0.35843625  0.48244361
#&gt;  [25]  0.53524041  0.18513068  2.31124663  1.16575986  0.32041542  3.50368377
#&gt;  [31] -1.00465442 -2.71547802 -1.84809109 -2.17684907 -0.55607988 -1.65445098
#&gt;  [37] -0.56980694 -0.56283147 -0.61697304 -1.68050494 -0.73657452 -1.11233661
#&gt;  [43] -0.86945280 -2.99812568 -1.57405028 -2.33964334 -1.86055039 -1.16168699
#&gt;  [49] -1.94460627 -2.66659373 -0.68291806 -0.04276485 -2.70999814  4.07998775
#&gt;  [55]  3.92608432 -3.23176985  1.16150604  0.36461222  1.37554546 -1.33991208
#&gt;  [61] -0.27327874  0.06561465 -0.46989453  2.57602440 -1.58258384  3.20820029
#&gt;  [67] -4.36535571 -0.49221693 -0.23172931  1.27279542  0.14340372 -0.63955336
#&gt;  [73] -2.48836915 -2.45402352 -1.99769497  0.85022079  0.97729907 -0.65016344
#&gt;  [79]  1.07879359  3.41875068 -1.10469580 -6.22321486  1.62403278 -1.96482688
#&gt;  [85]  0.18237357  1.50685612 -0.52711916 -2.77116144 -0.68528078 -0.50858194
#&gt;  [91]  1.32977539  1.18283073 -0.91275131  0.88646473 -2.67243128  1.74339325
#&gt;  [97]  0.85705167  1.58252482  1.05483796  0.98965858</code></pre>
<pre class="r"><code># x and y2 are highly correlated 
data.frame(x=x, y=y2) %&gt;%
        ggplot(aes(x=x, y =y2)) +
        geom_point() +
        geom_smooth(method=&#39;lm&#39;) +
        theme_minimal(base_size = 14)</code></pre>
<p><img src="/post/2025-05-22_partial_correlation_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<pre class="r"><code># x and y are not correlated
data.frame(x=x, y=y) %&gt;%
        ggplot(aes(x=x, y =y)) +
        geom_point() +
        geom_smooth(method=&#39;lm&#39;) +
        theme_minimal(base_size = 14)</code></pre>
<p><img src="/post/2025-05-22_partial_correlation_files/figure-html/unnamed-chunk-3-2.png" width="672" /></p>
<p>Let’s calculate the correlation and p-value</p>
<pre class="r"><code># use cor.test
a&lt;- cor.test(y, x, method = &quot;pearson&quot;) # Built-in

a </code></pre>
<pre><code>#&gt; 
#&gt; 	Pearson&#39;s product-moment correlation
#&gt; 
#&gt; data:  y and x
#&gt; t = 1.0303, df = 48, p-value = 0.308
#&gt; alternative hypothesis: true correlation is not equal to 0
#&gt; 95 percent confidence interval:
#&gt;  -0.1368577  0.4087073
#&gt; sample estimates:
#&gt;       cor 
#&gt; 0.1470934</code></pre>
<pre class="r"><code>tidy(a) %&gt;%
  kable(digits = 3, 
        caption = &quot;Correlation Test Results&quot;,
        col.names = c(&quot;Estimate&quot;, &quot;Statistic&quot;, &quot;p-value&quot;, &quot;Parameter&quot;, 
                     &quot;Lower CI&quot;, &quot;Upper CI&quot;, &quot;Method&quot;, &quot;Alternative&quot;))</code></pre>
<table>
<caption><span id="tab:unnamed-chunk-4">Table 1: </span>Correlation Test Results</caption>
<colgroup>
<col width="8%" />
<col width="9%" />
<col width="7%" />
<col width="9%" />
<col width="8%" />
<col width="8%" />
<col width="35%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th align="right">Estimate</th>
<th align="right">Statistic</th>
<th align="right">p-value</th>
<th align="right">Parameter</th>
<th align="right">Lower CI</th>
<th align="right">Upper CI</th>
<th align="left">Method</th>
<th align="left">Alternative</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">0.147</td>
<td align="right">1.03</td>
<td align="right">0.308</td>
<td align="right">48</td>
<td align="right">-0.137</td>
<td align="right">0.409</td>
<td align="left">Pearson’s product-moment correlation</td>
<td align="left">two.sided</td>
</tr>
</tbody>
</table>
<pre class="r"><code>## use linear model
b&lt;- lm(y ~ 1 + x) # Equivalent linear model: y = Beta0*1 + Beta1*x


# Create a nice table
tidy(b) %&gt;%
  kable(digits = 3, 
        caption = &quot;Linear Model Results&quot;,
        col.names = c(&quot;Term&quot;, &quot;Estimate&quot;, &quot;Std Error&quot;, &quot;t-statistic&quot;, &quot;p-value&quot;))</code></pre>
<table>
<caption><span id="tab:unnamed-chunk-4">Table 1: </span>Linear Model Results</caption>
<thead>
<tr class="header">
<th align="left">Term</th>
<th align="right">Estimate</th>
<th align="right">Std Error</th>
<th align="right">t-statistic</th>
<th align="right">p-value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">(Intercept)</td>
<td align="right">-0.159</td>
<td align="right">0.231</td>
<td align="right">-0.689</td>
<td align="right">0.494</td>
</tr>
<tr class="even">
<td align="left">x</td>
<td align="right">0.241</td>
<td align="right">0.233</td>
<td align="right">1.030</td>
<td align="right">0.308</td>
</tr>
</tbody>
</table>
<p>The p-value is the same (p = 0.308), but the coefficient is not (0.147 vs 0.241).</p>
<p>It turns out, we need to standardize x and y to get the same correlation coefficient.</p>
<pre class="r"><code># need to scale to get the same correlation coefficient
c&lt;- lm(scale(y) ~ 1 + scale(x))

tidy(c) %&gt;%
  kable(digits = 3, 
        caption = &quot;Linear Model Results&quot;,
        col.names = c(&quot;Term&quot;, &quot;Estimate&quot;, &quot;Std Error&quot;, &quot;t-statistic&quot;, &quot;p-value&quot;))</code></pre>
<table>
<caption><span id="tab:unnamed-chunk-5">Table 2: </span>Linear Model Results</caption>
<thead>
<tr class="header">
<th align="left">Term</th>
<th align="right">Estimate</th>
<th align="right">Std Error</th>
<th align="right">t-statistic</th>
<th align="right">p-value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">(Intercept)</td>
<td align="right">0.000</td>
<td align="right">0.141</td>
<td align="right">0.00</td>
<td align="right">1.000</td>
</tr>
<tr class="even">
<td align="left">scale(x)</td>
<td align="right">0.147</td>
<td align="right">0.143</td>
<td align="right">1.03</td>
<td align="right">0.308</td>
</tr>
</tbody>
</table>
<p>Now, we get the same 0.147 of the coefficient.</p>
</div>
<div id="why-need-to-standardize-it-to-get-the-right-correlation-coefficient" class="section level2">
<h2>Why need to standardize it to get the right correlation coefficient</h2>
<div id="pearsons-correlation-coefficient" class="section level3">
<h3>Pearson’s Correlation Coefficient</h3>
<p>The formula for Pearson’s correlation coefficient (<span class="math inline">\(r\)</span>) is:</p>
<p><span class="math display">\[
r = \frac{\sum_{i=1}^{n} (x_i - \bar{x})(y_i - \bar{y})}
         {\sqrt{\sum_{i=1}^{n} (x_i - \bar{x})^2} \sqrt{\sum_{i=1}^{n} (y_i - \bar{y})^2}}
\]</span></p>
</div>
<div id="regression-slope-for-standardized-variables" class="section level3">
<h3>Regression Slope for Standardized Variables</h3>
<p>When <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> are standardized to <span class="math inline">\(z\)</span>-scores:</p>
<p><span class="math display">\[
x_{std} = \frac{x - \bar{x}}{SD(x)}
\]</span></p>
<p><span class="math display">\[
y_{std} = \frac{y - \bar{y}}{SD(y)}
\]</span></p>
<p>The regression slope (<span class="math inline">\(\beta\)</span>) is:</p>
<p><span class="math display">\[
\beta = \frac{\sum_{i=1}^{n} x_{std,i} \, y_{std,i}}{\sum_{i=1}^{n} x_{std,i}^2}
\]</span></p>
</div>
<div id="proof-of-equivalence" class="section level3">
<h3>Proof of Equivalence</h3>
<p>Since standardized variables have <span class="math inline">\(SD(x_{std}) = SD(y_{std}) = 1\)</span>,</p>
<p><span class="math display">\[
\sum_{i=1}^{n} x_{std,i}^2 = n
\]</span></p>
<p>So,</p>
<p><span class="math display">\[
\beta = \frac{\sum_{i=1}^{n} x_{std,i} \, y_{std,i}}{n}
\]</span></p>
<p>Expanding <span class="math inline">\(x_{std}\)</span> and <span class="math inline">\(y_{std}\)</span>:</p>
<p><span class="math display">\[
\beta = \frac{1}{n} \cdot \frac{\sum_{i=1}^{n} (x_i - \bar{x})(y_i - \bar{y})}{SD(x) \cdot SD(y)}
\]</span></p>
<p>Therefore,</p>
<p><span class="math display">\[
\beta = r
\]</span></p>
<p>Now, let’s calculate the pairs of x and y1 which are significantly correlated:</p>
<pre class="r"><code>a2&lt;- cor.test(y2, x, method = &quot;pearson&quot;) # Built-in

a2</code></pre>
<pre><code>#&gt; 
#&gt; 	Pearson&#39;s product-moment correlation
#&gt; 
#&gt; data:  y2 and x
#&gt; t = 13.164, df = 48, p-value &lt; 2.2e-16
#&gt; alternative hypothesis: true correlation is not equal to 0
#&gt; 95 percent confidence interval:
#&gt;  0.8048173 0.9333683
#&gt; sample estimates:
#&gt;       cor 
#&gt; 0.8849249</code></pre>
<pre class="r"><code>tidy(a2) %&gt;%
  kable(digits = 3, 
        caption = &quot;Correlation Test Results&quot;,
        col.names = c(&quot;Estimate&quot;, &quot;Statistic&quot;, &quot;p-value&quot;, &quot;Parameter&quot;, 
                     &quot;Lower CI&quot;, &quot;Upper CI&quot;, &quot;Method&quot;, &quot;Alternative&quot;))</code></pre>
<table>
<caption><span id="tab:unnamed-chunk-6">Table 3: </span>Correlation Test Results</caption>
<colgroup>
<col width="8%" />
<col width="9%" />
<col width="7%" />
<col width="9%" />
<col width="8%" />
<col width="8%" />
<col width="35%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th align="right">Estimate</th>
<th align="right">Statistic</th>
<th align="right">p-value</th>
<th align="right">Parameter</th>
<th align="right">Lower CI</th>
<th align="right">Upper CI</th>
<th align="left">Method</th>
<th align="left">Alternative</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">0.885</td>
<td align="right">13.164</td>
<td align="right">0</td>
<td align="right">48</td>
<td align="right">0.805</td>
<td align="right">0.933</td>
<td align="left">Pearson’s product-moment correlation</td>
<td align="left">two.sided</td>
</tr>
</tbody>
</table>
<p>correlation of 0.885 with a p value of 0.</p>
<pre class="r"><code>## use linear model
b2&lt;- lm(y2 ~ 1 + x) # Equivalent linear model: y = Beta0*1 + Beta1*x


# Create a nice table
tidy(b2) %&gt;%
  kable(digits = 3, 
        caption = &quot;Linear Model Results&quot;,
        col.names = c(&quot;Term&quot;, &quot;Estimate&quot;, &quot;Std Error&quot;, &quot;t-statistic&quot;, &quot;p-value&quot;))</code></pre>
<table>
<caption><span id="tab:unnamed-chunk-7">Table 4: </span>Linear Model Results</caption>
<thead>
<tr class="header">
<th align="left">Term</th>
<th align="right">Estimate</th>
<th align="right">Std Error</th>
<th align="right">t-statistic</th>
<th align="right">p-value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">(Intercept)</td>
<td align="right">-0.072</td>
<td align="right">0.136</td>
<td align="right">-0.526</td>
<td align="right">0.601</td>
</tr>
<tr class="even">
<td align="left">x</td>
<td align="right">1.811</td>
<td align="right">0.138</td>
<td align="right">13.164</td>
<td align="right">0.000</td>
</tr>
</tbody>
</table>
<pre class="r"><code># need to scale to get the same correlation coefficient
c2&lt;- lm(scale(y2) ~ 1 + scale(x))

tidy(c2) %&gt;%
  kable(digits = 3, 
        caption = &quot;Linear Model Results&quot;,
        col.names = c(&quot;Term&quot;, &quot;Estimate&quot;, &quot;Std Error&quot;, &quot;t-statistic&quot;, &quot;p-value&quot;))</code></pre>
<table>
<caption><span id="tab:unnamed-chunk-8">Table 5: </span>Linear Model Results</caption>
<thead>
<tr class="header">
<th align="left">Term</th>
<th align="right">Estimate</th>
<th align="right">Std Error</th>
<th align="right">t-statistic</th>
<th align="right">p-value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">(Intercept)</td>
<td align="right">0.000</td>
<td align="right">0.067</td>
<td align="right">0.000</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">scale(x)</td>
<td align="right">0.885</td>
<td align="right">0.067</td>
<td align="right">13.164</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>we get the same coefficient of <code>0.885</code> and p-value of 0.</p>
<p>It is very satisfying to see we get the same results using different methods.</p>
</div>
</div>
<div id="a-practical-example-in-calculating-partial-correlation" class="section level2">
<h2>A practical example in calculating partial correlation</h2>
<p>Go to <a href="https://depmap.org/portal/data_page/?tab=currentRelease">Depmap</a>
Download the CRISPR screening dependency data <code>CRISPRGeneEffect.csv</code></p>
<p>The <code>CRISPRGeneEffect.csv</code> file from DepMap contains results from genome-wide CRISPR-Cas9 knockout screens across hundreds of cancer cell lines, reporting how essential each gene is for cell survival. The gene effect scores, calculated using the CERES or Chronos algorithms, indicate the impact of knocking out each gene: <strong>lower scores mean a gene is more essential for that cell line’s viability</strong>.</p>
<p>This dataset enables researchers to identify cancer-specific genetic vulnerabilities and potential therapeutic targets by showing which genes are critical for the survival of different cancer types.</p>
<p>The file is &gt; 400MB.</p>
<p>Also download the <code>Model.csv</code> file which contains the metadata information (e.g, cancer type for each cell line)</p>
<pre class="r"><code>library(readr)
library(dplyr)

#read in the data
crispr_score&lt;- read_csv(&quot;~/blog_data/CRISPRGeneEffect.csv&quot;)

crispr_score[1:5, 1:5]</code></pre>
<pre><code>#&gt; # A tibble: 5 × 5
#&gt;   ...1       `A1BG (1)` `A1CF (29974)` `A2M (2)` `A2ML1 (144568)`
#&gt;   &lt;chr&gt;           &lt;dbl&gt;          &lt;dbl&gt;     &lt;dbl&gt;            &lt;dbl&gt;
#&gt; 1 ACH-000001    -0.122         0.0426     0.0261          -0.148 
#&gt; 2 ACH-000004    -0.0165       -0.0885    -0.0882          -0.0412
#&gt; 3 ACH-000005    -0.185         0.00330    0.161            0.0862
#&gt; 4 ACH-000007    -0.0719       -0.114      0.0829           0.0996
#&gt; 5 ACH-000009    -0.0192       -0.135      0.0603           0.0766</code></pre>
<p>We need to clean up the column names. remove the parentheses and the ENTREZE ID (numbers).</p>
<p>NOTE: this type of regular expression is a perfect question for <code>LLM</code>.</p>
<pre class="r"><code>crispr_score&lt;- crispr_score %&gt;%
        dplyr::rename(ModelID = `...1`) %&gt;%
        rename_with(~str_trim(str_remove(.x, &quot; \\(.*\\)$&quot;)), -1)

crispr_score[1:5, 1:5]</code></pre>
<pre><code>#&gt; # A tibble: 5 × 5
#&gt;   ModelID       A1BG     A1CF     A2M   A2ML1
#&gt;   &lt;chr&gt;        &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
#&gt; 1 ACH-000001 -0.122   0.0426   0.0261 -0.148 
#&gt; 2 ACH-000004 -0.0165 -0.0885  -0.0882 -0.0412
#&gt; 3 ACH-000005 -0.185   0.00330  0.161   0.0862
#&gt; 4 ACH-000007 -0.0719 -0.114    0.0829  0.0996
#&gt; 5 ACH-000009 -0.0192 -0.135    0.0603  0.0766</code></pre>
<pre class="r"><code>meta&lt;- read_csv(&quot;~/blog_data/Model.csv&quot;)
head(meta)</code></pre>
<pre><code>#&gt; # A tibble: 6 × 47
#&gt;   ModelID    PatientID CellLineName StrippedCellLineName DepmapModelType
#&gt;   &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;                &lt;chr&gt;          
#&gt; 1 ACH-000001 PT-gj46wT NIH:OVCAR-3  NIHOVCAR3            HGSOC          
#&gt; 2 ACH-000002 PT-5qa3uk HL-60        HL60                 AML            
#&gt; 3 ACH-000003 PT-puKIyc CACO2        CACO2                COAD           
#&gt; 4 ACH-000004 PT-q4K2cp HEL          HEL                  AML            
#&gt; 5 ACH-000005 PT-q4K2cp HEL 92.1.7   HEL9217              AML            
#&gt; 6 ACH-000006 PT-ej13Dz MONO-MAC-6   MONOMAC6             AMOL           
#&gt; # ℹ 42 more variables: OncotreeLineage &lt;chr&gt;, OncotreePrimaryDisease &lt;chr&gt;,
#&gt; #   OncotreeSubtype &lt;chr&gt;, OncotreeCode &lt;chr&gt;, PatientSubtypeFeatures &lt;chr&gt;,
#&gt; #   RRID &lt;chr&gt;, Age &lt;dbl&gt;, AgeCategory &lt;chr&gt;, Sex &lt;chr&gt;, PatientRace &lt;chr&gt;,
#&gt; #   PrimaryOrMetastasis &lt;chr&gt;, SampleCollectionSite &lt;chr&gt;, SourceType &lt;chr&gt;,
#&gt; #   SourceDetail &lt;chr&gt;, CatalogNumber &lt;chr&gt;, ModelType &lt;chr&gt;,
#&gt; #   TissueOrigin &lt;lgl&gt;, ModelDerivationMaterial &lt;chr&gt;, ModelTreatment &lt;chr&gt;,
#&gt; #   PatientTreatmentStatus &lt;chr&gt;, PatientTreatmentType &lt;chr&gt;, …</code></pre>
<pre class="r"><code>table(meta$OncotreeLineage)</code></pre>
<pre><code>#&gt; 
#&gt;             Adrenal Gland          Ampulla of Vater             B lymphoblast 
#&gt;                         2                         4                         1 
#&gt;             Biliary Tract     Bladder/Urinary Tract                      Bone 
#&gt;                        46                        39                        90 
#&gt;                     Bowel                    Breast                    Cervix 
#&gt;                        99                        96                        25 
#&gt;                 CNS/Brain                 Embryonal         Esophagus/Stomach 
#&gt;                       125                         1                       103 
#&gt;                       Eye                Fibroblast                      Hair 
#&gt;                        24                        42                         2 
#&gt;             Head and Neck                    Kidney                     Liver 
#&gt;                        95                        73                        29 
#&gt;                      Lung                  Lymphoid                    Muscle 
#&gt;                       260                       260                         5 
#&gt;                   Myeloid                    Normal                     Other 
#&gt;                        88                        13                         2 
#&gt;      Ovary/Fallopian Tube                  Pancreas Peripheral Nervous System 
#&gt;                        75                        68                        60 
#&gt;                    Pleura                  Prostate                      Skin 
#&gt;                        36                        15                       149 
#&gt;               Soft Tissue                    Testis                   Thyroid 
#&gt;                        86                        12                        25 
#&gt;                    Uterus              Vulva/Vagina 
#&gt;                        44                         5</code></pre>
<pre class="r"><code># subset only the breast cancer cell line
breast_meta&lt;- meta %&gt;%
        select(ModelID, OncotreeLineage) %&gt;%
        mutate(breast = case_when(
                OncotreeLineage == &quot;Breast&quot; ~ &quot;yes&quot;,
                TRUE ~ &quot;no&quot;
        ))


crispr_all&lt;- inner_join(meta, crispr_score)


crispr_all&lt;- crispr_all %&gt;%
        mutate(breast = case_when(
                OncotreeLineage == &quot;Breast&quot; ~ &quot;yes&quot;,
                TRUE ~ &quot;no&quot;
        ))



ggplot(crispr_all,  aes(x= FOXA1, y= ESR1)) +
        geom_point(aes(color = breast)) +
        geom_smooth(method=&#39;lm&#39;, formula= y~x) +
        facet_wrap(~breast) </code></pre>
<p><img src="/post/2025-05-22_partial_correlation_files/figure-html/unnamed-chunk-11-1.png" width="672" />
It looks like the FOXA1 and ESR1 CRISPR dependency score are more correlated in Breast cancer.</p>
<p>Let’s calculate the correlation and p-value</p>
<pre class="r"><code>cor.test(crispr_all$FOXA1[crispr_all$breast == &quot;yes&quot;], crispr_all$ESR1[crispr_all$breast == &quot;yes&quot;])</code></pre>
<pre><code>#&gt; 
#&gt; 	Pearson&#39;s product-moment correlation
#&gt; 
#&gt; data:  crispr_all$FOXA1[crispr_all$breast == &quot;yes&quot;] and crispr_all$ESR1[crispr_all$breast == &quot;yes&quot;]
#&gt; t = 4.3022, df = 51, p-value = 7.658e-05
#&gt; alternative hypothesis: true correlation is not equal to 0
#&gt; 95 percent confidence interval:
#&gt;  0.2855579 0.6900675
#&gt; sample estimates:
#&gt;       cor 
#&gt; 0.5160227</code></pre>
<pre class="r"><code>cor.test(crispr_all$FOXA1, crispr_all$ESR1)</code></pre>
<pre><code>#&gt; 
#&gt; 	Pearson&#39;s product-moment correlation
#&gt; 
#&gt; data:  crispr_all$FOXA1 and crispr_all$ESR1
#&gt; t = 14.076, df = 1176, p-value &lt; 2.2e-16
#&gt; alternative hypothesis: true correlation is not equal to 0
#&gt; 95 percent confidence interval:
#&gt;  0.3297560 0.4275629
#&gt; sample estimates:
#&gt;       cor 
#&gt; 0.3797201</code></pre>
<ul>
<li><p>All cell lines (r = 0.38): This includes the confounding effect of cancer type. Different cancer types have different baseline dependencies for both FOXA1 and ESR1.</p></li>
<li><p>Breast cancer only (r = 0.52): This removes the cancer type confounding, showing the “true” relationship within a homogeneous cancer type.</p></li>
<li><p>The increase from 0.38 to 0.52 suggests that cancer type was acting as a confounding variable.</p></li>
</ul>
<div id="lets-use-linear-model-to-calculate-correlation" class="section level3">
<h3>Let’s use linear model to calculate correlation</h3>
<pre class="r"><code># need to scale to get the same correlation coefficient
lm_cor &lt;- lm(scale(crispr_all$ESR1) ~ 1 + scale(crispr_all$FOXA1))

summary(lm_cor)</code></pre>
<pre><code>#&gt; 
#&gt; Call:
#&gt; lm(formula = scale(crispr_all$ESR1) ~ 1 + scale(crispr_all$FOXA1))
#&gt; 
#&gt; Residuals:
#&gt;     Min      1Q  Median      3Q     Max 
#&gt; -7.7928 -0.3632  0.0598  0.4546  3.5335 
#&gt; 
#&gt; Coefficients:
#&gt;                          Estimate Std. Error t value Pr(&gt;|t|)    
#&gt; (Intercept)             4.703e-17  2.697e-02    0.00        1    
#&gt; scale(crispr_all$FOXA1) 3.797e-01  2.698e-02   14.08   &lt;2e-16 ***
#&gt; ---
#&gt; Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
#&gt; 
#&gt; Residual standard error: 0.9255 on 1176 degrees of freedom
#&gt; Multiple R-squared:  0.1442,	Adjusted R-squared:  0.1435 
#&gt; F-statistic: 198.1 on 1 and 1176 DF,  p-value: &lt; 2.2e-16</code></pre>
<pre class="r"><code># a nice table
tidy(lm_cor) %&gt;%
  kable(digits = 3, 
        caption = &quot;Linear Model Results&quot;,
        col.names = c(&quot;Term&quot;, &quot;Estimate&quot;, &quot;Std Error&quot;, &quot;t-statistic&quot;, &quot;p-value&quot;))</code></pre>
<table>
<caption><span id="tab:unnamed-chunk-13">Table 6: </span>Linear Model Results</caption>
<thead>
<tr class="header">
<th align="left">Term</th>
<th align="right">Estimate</th>
<th align="right">Std Error</th>
<th align="right">t-statistic</th>
<th align="right">p-value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">(Intercept)</td>
<td align="right">0.00</td>
<td align="right">0.027</td>
<td align="right">0.000</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">scale(crispr_all$FOXA1)</td>
<td align="right">0.38</td>
<td align="right">0.027</td>
<td align="right">14.076</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>The results is the same as <code>cor.test(crispr_all$FOXA1, crispr_all$ESR1)</code></p>
</div>
</div>
<div id="adding-the-cancer-type-as-a-covariate-to-calculate-partial-correlation" class="section level2">
<h2>adding the cancer type as a covariate to calculate partial correlation</h2>
<pre class="r"><code>lm_cor_partial &lt;- lm(scale(crispr_all$ESR1) ~ scale(crispr_all$FOXA1) + crispr_all$OncotreeLineage)

unique(crispr_all$OncotreeLineage) %&gt;% sort()</code></pre>
<pre><code>#&gt;  [1] &quot;Adrenal Gland&quot;             &quot;Ampulla of Vater&quot;         
#&gt;  [3] &quot;Biliary Tract&quot;             &quot;Bladder/Urinary Tract&quot;    
#&gt;  [5] &quot;Bone&quot;                      &quot;Bowel&quot;                    
#&gt;  [7] &quot;Breast&quot;                    &quot;Cervix&quot;                   
#&gt;  [9] &quot;CNS/Brain&quot;                 &quot;Esophagus/Stomach&quot;        
#&gt; [11] &quot;Eye&quot;                       &quot;Fibroblast&quot;               
#&gt; [13] &quot;Head and Neck&quot;             &quot;Kidney&quot;                   
#&gt; [15] &quot;Liver&quot;                     &quot;Lung&quot;                     
#&gt; [17] &quot;Lymphoid&quot;                  &quot;Myeloid&quot;                  
#&gt; [19] &quot;Other&quot;                     &quot;Ovary/Fallopian Tube&quot;     
#&gt; [21] &quot;Pancreas&quot;                  &quot;Peripheral Nervous System&quot;
#&gt; [23] &quot;Pleura&quot;                    &quot;Prostate&quot;                 
#&gt; [25] &quot;Skin&quot;                      &quot;Soft Tissue&quot;              
#&gt; [27] &quot;Testis&quot;                    &quot;Thyroid&quot;                  
#&gt; [29] &quot;Uterus&quot;                    &quot;Vulva/Vagina&quot;</code></pre>
<p>We have 30 lineages, and by default the reference group is <code>Adrenal Gland</code> which is the first sorted alphabetically.</p>
<pre class="r"><code>summary(lm_cor_partial)</code></pre>
<pre><code>#&gt; 
#&gt; Call:
#&gt; lm(formula = scale(crispr_all$ESR1) ~ scale(crispr_all$FOXA1) + 
#&gt;     crispr_all$OncotreeLineage)
#&gt; 
#&gt; Residuals:
#&gt;     Min      1Q  Median      3Q     Max 
#&gt; -7.1693 -0.3711  0.0329  0.4560  3.5215 
#&gt; 
#&gt; Coefficients:
#&gt;                                                     Estimate Std. Error t value
#&gt; (Intercept)                                           1.2382     0.9029   1.371
#&gt; scale(crispr_all$FOXA1)                               0.3021     0.0299  10.104
#&gt; crispr_all$OncotreeLineageAmpulla of Vater           -0.9696     1.0094  -0.961
#&gt; crispr_all$OncotreeLineageBiliary Tract              -1.2050     0.9157  -1.316
#&gt; crispr_all$OncotreeLineageBladder/Urinary Tract      -1.0857     0.9161  -1.185
#&gt; crispr_all$OncotreeLineageBone                       -1.1071     0.9126  -1.213
#&gt; crispr_all$OncotreeLineageBowel                      -1.0521     0.9100  -1.156
#&gt; crispr_all$OncotreeLineageBreast                     -2.1645     0.9131  -2.371
#&gt; crispr_all$OncotreeLineageCervix                     -0.8041     0.9276  -0.867
#&gt; crispr_all$OncotreeLineageCNS/Brain                  -1.3465     0.9079  -1.483
#&gt; crispr_all$OncotreeLineageEsophagus/Stomach          -1.1176     0.9094  -1.229
#&gt; crispr_all$OncotreeLineageEye                        -1.3322     0.9325  -1.429
#&gt; crispr_all$OncotreeLineageFibroblast                 -0.8213     1.2769  -0.643
#&gt; crispr_all$OncotreeLineageHead and Neck              -1.1233     0.9087  -1.236
#&gt; crispr_all$OncotreeLineageKidney                     -1.4100     0.9161  -1.539
#&gt; crispr_all$OncotreeLineageLiver                      -1.0757     0.9215  -1.167
#&gt; crispr_all$OncotreeLineageLung                       -1.2399     0.9064  -1.368
#&gt; crispr_all$OncotreeLineageLymphoid                   -1.1707     0.9078  -1.290
#&gt; crispr_all$OncotreeLineageMyeloid                    -1.1257     0.9136  -1.232
#&gt; crispr_all$OncotreeLineageOther                      -0.7691     1.2769  -0.602
#&gt; crispr_all$OncotreeLineageOvary/Fallopian Tube       -1.6780     0.9105  -1.843
#&gt; crispr_all$OncotreeLineagePancreas                   -1.1337     0.9124  -1.242
#&gt; crispr_all$OncotreeLineagePeripheral Nervous System  -1.1941     0.9131  -1.308
#&gt; crispr_all$OncotreeLineagePleura                     -1.0867     0.9241  -1.176
#&gt; crispr_all$OncotreeLineageProstate                   -0.8901     0.9481  -0.939
#&gt; crispr_all$OncotreeLineageSkin                       -1.0486     0.9090  -1.154
#&gt; crispr_all$OncotreeLineageSoft Tissue                -1.2275     0.9129  -1.345
#&gt; crispr_all$OncotreeLineageTestis                     -1.9009     0.9891  -1.922
#&gt; crispr_all$OncotreeLineageThyroid                    -1.1388     0.9430  -1.208
#&gt; crispr_all$OncotreeLineageUterus                     -1.2534     0.9161  -1.368
#&gt; crispr_all$OncotreeLineageVulva/Vagina               -1.3165     1.1058  -1.191
#&gt;                                                     Pr(&gt;|t|)    
#&gt; (Intercept)                                           0.1705    
#&gt; scale(crispr_all$FOXA1)                               &lt;2e-16 ***
#&gt; crispr_all$OncotreeLineageAmpulla of Vater            0.3370    
#&gt; crispr_all$OncotreeLineageBiliary Tract               0.1884    
#&gt; crispr_all$OncotreeLineageBladder/Urinary Tract       0.2362    
#&gt; crispr_all$OncotreeLineageBone                        0.2254    
#&gt; crispr_all$OncotreeLineageBowel                       0.2479    
#&gt; crispr_all$OncotreeLineageBreast                      0.0179 *  
#&gt; crispr_all$OncotreeLineageCervix                      0.3862    
#&gt; crispr_all$OncotreeLineageCNS/Brain                   0.1383    
#&gt; crispr_all$OncotreeLineageEsophagus/Stomach           0.2193    
#&gt; crispr_all$OncotreeLineageEye                         0.1534    
#&gt; crispr_all$OncotreeLineageFibroblast                  0.5203    
#&gt; crispr_all$OncotreeLineageHead and Neck               0.2167    
#&gt; crispr_all$OncotreeLineageKidney                      0.1240    
#&gt; crispr_all$OncotreeLineageLiver                       0.2433    
#&gt; crispr_all$OncotreeLineageLung                        0.1716    
#&gt; crispr_all$OncotreeLineageLymphoid                    0.1974    
#&gt; crispr_all$OncotreeLineageMyeloid                     0.2182    
#&gt; crispr_all$OncotreeLineageOther                       0.5471    
#&gt; crispr_all$OncotreeLineageOvary/Fallopian Tube        0.0656 .  
#&gt; crispr_all$OncotreeLineagePancreas                    0.2143    
#&gt; crispr_all$OncotreeLineagePeripheral Nervous System   0.1912    
#&gt; crispr_all$OncotreeLineagePleura                      0.2399    
#&gt; crispr_all$OncotreeLineageProstate                    0.3481    
#&gt; crispr_all$OncotreeLineageSkin                        0.2489    
#&gt; crispr_all$OncotreeLineageSoft Tissue                 0.1790    
#&gt; crispr_all$OncotreeLineageTestis                      0.0549 .  
#&gt; crispr_all$OncotreeLineageThyroid                     0.2274    
#&gt; crispr_all$OncotreeLineageUterus                      0.1715    
#&gt; crispr_all$OncotreeLineageVulva/Vagina                0.2341    
#&gt; ---
#&gt; Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
#&gt; 
#&gt; Residual standard error: 0.9029 on 1147 degrees of freedom
#&gt; Multiple R-squared:  0.2056,	Adjusted R-squared:  0.1848 
#&gt; F-statistic: 9.896 on 30 and 1147 DF,  p-value: &lt; 2.2e-16</code></pre>
<p>Interpreting the linear model coefficients</p>
<ul>
<li>FOXA1 coefficient (scale(crispr_all$FOXA1) = 0.3021)</li>
</ul>
<p>Meaning: For every 1 standard deviation increase in FOXA1 dependency score, ESR1 dependency score increases by 0.3021 standard deviations, controlling for cancer type.</p>
<p>P-value &lt; 2e-16: Highly significant relationship
This 0.3021 is your partial correlation coefficient between FOXA1 and ESR1, controlling for cancer type.</p>
<ul>
<li>Cancer type coefficients (e.g., Breast = -2.1645)</li>
</ul>
<p>Meaning: Each coefficient represents the difference in ESR1 dependency (in standard deviations) between that cancer type and the reference cancer type (Adrenal Gland).</p>
<p>Breast coefficient (-2.1645, p = 0.0179): Breast cancer cell lines have significantly lower ESR1 dependency scores (lower means more dependent) compared to the reference cancer type (Adrenal Gland), on average.</p>
<p>Most other cancer types: Non-significant differences from the reference type.</p>
<p>Biological interpretation:</p>
<p>FOXA1 and ESR1 have a moderate positive partial correlation (0.30) across cancer types.</p>
<p>This suggests these genes may be part of related pathways or have synthetic lethal relationships.</p>
<p>Breast cancer shows even stronger co-dependency (0.53), which makes biological sense given the importance of estrogen signaling in breast cancer.</p>
<p>The significant negative coefficient for breast cancer (-2.16) indicates breast cancers generally have lower ESR1 dependency scores overall.</p>
</div>
<div id="what-we-found-suggest" class="section level2">
<h2>What we found suggest</h2>
<ul>
<li><p>FOXA1-ESR1 have genuine co-dependency (partial r = 0.30).</p></li>
<li><p>Breast cancer amplifies this relationship (r = 0.53 within breast).</p></li>
<li><p>Cancer type was indeed confounding the raw correlation (0.37).</p></li>
</ul>
<p>The relationship is biologically meaningful across cancer types, but particularly strong in breast cancer.</p>
</div>
<div id="other-usages-of-partial-correlation" class="section level2">
<h2>Other usages of partial correlation</h2>
<p>Partial Correlation Improves Network Accuracy</p>
<ol style="list-style-type: decimal">
<li>Eliminating Spurious Correlations
Problem: Standard Pearson/Spearman correlations detect both direct interactions and indirect relationships mediated by shared regulators or technical confounders (e.g., batch effects).</li>
</ol>
<p>Solution: Partial correlation removes the linear effects of all other variables, revealing direct dependencies between gene pairs.</p>
<ol start="2" style="list-style-type: decimal">
<li>Biological Validation
Studies show partial correlation networks:</li>
</ol>
<p>Reduce false positives by 30-50% compared to correlation networks in cancer genomics
Align better with experimentally validated interactions (e.g., ChIP-seq/TF binding data).</p>
<p>Further readings:</p>
<ul>
<li><a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC3332147/">Partial correlation network analyses to detect altered gene interactions in human disease: using preeclampsia as a model</a></li>
<li><a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC4194134/">Biological network inference using low order partial correlation</a></li>
</ul>
</div>

    </div>

    


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/bioinformatics/">bioinformatics</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/depmap/">Depmap</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/crispr/">CRISPR</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/data/">data</a>
  
</div>




    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/pca-tcga2/">PCA analysis on TCGA bulk RNAseq data continued</a></li>
        
        <li><a href="/post/pca-scatac/">PCA analysis on scATACseq data</a></li>
        
        <li><a href="/post/pca-tcga/">PCA analysis on TCGA bulk RNAseq data</a></li>
        
        <li><a href="/post/biotech-data/">Biotech Data Strategy: Building a Scalable Foundation for Startups</a></li>
        
        <li><a href="/post/review-2024/">Review 2024</a></li>
        
      </ul>
    </div>
    

    
    <div class="article-widget">
      <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/post/cancer-drug/" rel="next">How Cancer Drugs Really Work</a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/post/multiomics-mofa2/" rel="prev">Multi-Omics Integration Strategy and Deep Diving into MOFA2</a>
  </div>
  
</div>

    </div>
    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "divingintogeneticsandgenomics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming &lsquo;Tommy&rsquo; Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

