<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.42" />
  <meta name="author" content="Ming Tang">

  
  
  
  
    
      
    
  
  <meta name="description" content="Objective Enhancer aberrations are beginning to emerge as a key epigenetic feature of colorectal cancers (CRC), however, a comprehensive knowledge of chromatin state patterns in tumour progression, heterogeneity of these patterns and imparted therapeutic opportunities remain poorly described. Design We performed comprehensive epigenomic characterisation by mapping 222 chromatin profiles from 69 samples (33 colorectal adenocarcinomas, 4 adenomas, 21 matched normal tissues and 11 colon cancer cell lines) for six histone modification marks: H3K4me3 for Pol II-bound and CpG-rich promoters, H3K4me1 for poised enhancers, H3K27ac for enhancers and transcriptionally active promoters, H3K79me2 for transcribed regions, H3K27me3 for polycomb repressed regions and H3K9me3 for heterochromatin.Results We demonstrate that H3K27ac-marked active enhancer state could distinguish between different stages of CRC progression. By epigenomic editing, we present evidence that gains of tumour-specific enhancers for crucial oncogenes, such as ASCL2 and FZD10, was required for excessive proliferation. Consistently, combination of MEK plus bromodomain inhibition was found to have synergistic effects in CRC patient-derived xenograft models. Probing intertumour heterogeneity, we identified four distinct enhancer subtypes (EPIgenome-based Classification, EpiC), three of which correlate well with previously defined transcriptomic subtypes (consensus molecular subtypes, CMSs). Importantly, CMS2 can be divided into two EpiC subgroups with significant survival differences. Leveraging such correlation, we devised a combinatorial therapeutic strategy of enhancer-blocking bromodomain inhibitors with pathway-specific inhibitors (PARPi, EGFRi, TGFÎ²i, mTORi and SRCi) for EpiC groups.Conclusion Our data suggest that the dynamics of active enhancer underlies CRC progression and the patient-specific enhancer patterns can be leveraged for precision combination therapy.">

  
  <link rel="alternate" hreflang="en-us" href="/publication/2021-05-30-chromatin-state-colorectal-cancer/">

  


  

  
  
  <meta name="theme-color" content="#328cc1">
  
  
  
  
    
  
  
    
    
      
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
      
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-84019592-2', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="DNA confesses Data speak">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/publication/2021-05-30-chromatin-state-colorectal-cancer/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@https://twitter.com/tangming2005">
  <meta property="twitter:creator" content="@https://twitter.com/tangming2005">
  
  <meta property="og:site_name" content="DNA confesses Data speak">
  <meta property="og:url" content="/publication/2021-05-30-chromatin-state-colorectal-cancer/">
  <meta property="og:title" content="Chromatin state dynamics confers specific therapeutic strategies in enhancer subtypes of colorectal cancer | DNA confesses Data speak">
  <meta property="og:description" content="Objective Enhancer aberrations are beginning to emerge as a key epigenetic feature of colorectal cancers (CRC), however, a comprehensive knowledge of chromatin state patterns in tumour progression, heterogeneity of these patterns and imparted therapeutic opportunities remain poorly described. Design We performed comprehensive epigenomic characterisation by mapping 222 chromatin profiles from 69 samples (33 colorectal adenocarcinomas, 4 adenomas, 21 matched normal tissues and 11 colon cancer cell lines) for six histone modification marks: H3K4me3 for Pol II-bound and CpG-rich promoters, H3K4me1 for poised enhancers, H3K27ac for enhancers and transcriptionally active promoters, H3K79me2 for transcribed regions, H3K27me3 for polycomb repressed regions and H3K9me3 for heterochromatin.Results We demonstrate that H3K27ac-marked active enhancer state could distinguish between different stages of CRC progression. By epigenomic editing, we present evidence that gains of tumour-specific enhancers for crucial oncogenes, such as ASCL2 and FZD10, was required for excessive proliferation. Consistently, combination of MEK plus bromodomain inhibition was found to have synergistic effects in CRC patient-derived xenograft models. Probing intertumour heterogeneity, we identified four distinct enhancer subtypes (EPIgenome-based Classification, EpiC), three of which correlate well with previously defined transcriptomic subtypes (consensus molecular subtypes, CMSs). Importantly, CMS2 can be divided into two EpiC subgroups with significant survival differences. Leveraging such correlation, we devised a combinatorial therapeutic strategy of enhancer-blocking bromodomain inhibitors with pathway-specific inhibitors (PARPi, EGFRi, TGFÎ²i, mTORi and SRCi) for EpiC groups.Conclusion Our data suggest that the dynamics of active enhancer underlies CRC progression and the patient-specific enhancer patterns can be leveraged for precision combination therapy."><meta property="og:image" content="/img/headers/CRC_chromatin_state.png">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2021-05-30T00:00:00-04:00">
  
  <meta property="article:modified_time" content="2021-05-30T00:00:00-04:00">
  

  
  

  <title>Chromatin state dynamics confers specific therapeutic strategies in enhancer subtypes of colorectal cancer | DNA confesses Data speak</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">DNA confesses Data speak</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#talks">
            
            <span>Talks &amp; Teachings</span>
            
          </a>
        </li>

        
        

        
        
        
        
        

        <li class="nav-item">
          <a href="/#cv">
            
            <span>CV</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>

<div class="pub" itemscope itemtype="http://schema.org/CreativeWork">

  
<div class="article-header">
  
  
    <img src="/img/headers/CRC_chromatin_state.png" class="article-banner" itemprop="image">
  

  <span class="article-header-caption">My caption ðŸ˜„</span>
</div>



  <div class="article-container">
    <h1 itemprop="name">Chromatin state dynamics confers specific therapeutic strategies in enhancer subtypes of colorectal cancer</h1>
    <span class="pub-authors" itemprop="author">
      
      Elias Orouji, Ayush T Raman, &hellip;, <strong>Ming Tang</strong> et.al
      
    </span>
    <span class="pull-right">
      
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=Chromatin%20state%20dynamics%20confers%20specific%20therapeutic%20strategies%20in%20enhancer%20subtypes%20of%20colorectal%20cancer&amp;url=%2fpublication%2f2021-05-30-chromatin-state-colorectal-cancer%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpublication%2f2021-05-30-chromatin-state-colorectal-cancer%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpublication%2f2021-05-30-chromatin-state-colorectal-cancer%2f&amp;title=Chromatin%20state%20dynamics%20confers%20specific%20therapeutic%20strategies%20in%20enhancer%20subtypes%20of%20colorectal%20cancer"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpublication%2f2021-05-30-chromatin-state-colorectal-cancer%2f&amp;title=Chromatin%20state%20dynamics%20confers%20specific%20therapeutic%20strategies%20in%20enhancer%20subtypes%20of%20colorectal%20cancer"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=Chromatin%20state%20dynamics%20confers%20specific%20therapeutic%20strategies%20in%20enhancer%20subtypes%20of%20colorectal%20cancer&amp;body=%2fpublication%2f2021-05-30-chromatin-state-colorectal-cancer%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


    </span>

    

    
    <h3>Abstract</h3>
    <p class="pub-abstract" itemprop="text">Objective Enhancer aberrations are beginning to emerge as a key epigenetic feature of colorectal cancers (CRC), however, a comprehensive knowledge of chromatin state patterns in tumour progression, heterogeneity of these patterns and imparted therapeutic opportunities remain poorly described. Design We performed comprehensive epigenomic characterisation by mapping 222 chromatin profiles from 69 samples (33 colorectal adenocarcinomas, 4 adenomas, 21 matched normal tissues and 11 colon cancer cell lines) for six histone modification marks: H3K4me3 for Pol II-bound and CpG-rich promoters, H3K4me1 for poised enhancers, H3K27ac for enhancers and transcriptionally active promoters, H3K79me2 for transcribed regions, H3K27me3 for polycomb repressed regions and H3K9me3 for heterochromatin.Results We demonstrate that H3K27ac-marked active enhancer state could distinguish between different stages of CRC progression. By epigenomic editing, we present evidence that gains of tumour-specific enhancers for crucial oncogenes, such as ASCL2 and FZD10, was required for excessive proliferation. Consistently, combination of MEK plus bromodomain inhibition was found to have synergistic effects in CRC patient-derived xenograft models. Probing intertumour heterogeneity, we identified four distinct enhancer subtypes (EPIgenome-based Classification, EpiC), three of which correlate well with previously defined transcriptomic subtypes (consensus molecular subtypes, CMSs). Importantly, CMS2 can be divided into two EpiC subgroups with significant survival differences. Leveraging such correlation, we devised a combinatorial therapeutic strategy of enhancer-blocking bromodomain inhibitors with pathway-specific inhibitors (PARPi, EGFRi, TGFÎ²i, mTORi and SRCi) for EpiC groups.Conclusion Our data suggest that the dynamics of active enhancer underlies CRC progression and the patient-specific enhancer patterns can be leveraged for precision combination therapy.</p>
    

    
    <div class="row">
      <div class="col-sm-1"></div>
      <div class="col-sm-10">
        <div class="row">
          <div class="col-xs-12 col-sm-3 pub-row-heading">Type</div>
          <div class="col-xs-12 col-sm-9">
            
            <a href="/publication/#2">
              Journal article
            </a>
            
          </div>
        </div>
      </div>
      <div class="col-sm-1"></div>
    </div>
    <div class="visible-xs space-below"></div>
    

    
    <div class="row">
      <div class="col-sm-1"></div>
      <div class="col-sm-10">
        <div class="row">
          <div class="col-xs-12 col-sm-3 pub-row-heading">Publication</div>
          <div class="col-xs-12 col-sm-9">In <em>Gut</em>, Gut.</div>
        </div>
      </div>
      <div class="col-sm-1"></div>
    </div>
    <div class="visible-xs space-below"></div>
    

    <div class="row">
      <div class="col-sm-1"></div>
      <div class="col-sm-10">
        <div class="row">
          <div class="col-xs-12 col-sm-3 pub-row-heading">Date</div>
          <div class="col-xs-12 col-sm-9" itemprop="datePublished">
            May, 2021
          </div>
        </div>
      </div>
      <div class="col-sm-1"></div>
    </div>
    <div class="visible-xs space-below"></div>

    <div class="row" style="padding-top: 10px">
      <div class="col-sm-1"></div>
      <div class="col-sm-10">
        <div class="row">
          <div class="col-xs-12 col-sm-3 pub-row-heading" style="line-height:34px;">Links</div>
          <div class="col-xs-12 col-sm-9">

            




<a class="btn btn-primary btn-outline" href="/files/papers/CRC_chromatin_state.pdf" target="_blank" rel="noopener">
  PDF
</a>
















          </div>
        </div>
      </div>
      <div class="col-sm-1"></div>
    </div>
    <div class="visible-xs space-below"></div>

    <div class="space-below"></div>

    <div class="article-style"><p>More detail can easily be written here using <em>Markdown</em> and $\rm \LaTeX$ math code.</p>
</div>

    


  </div>
</div>


<div class="article-container article-widget">
  <div class="post-nav">
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/publication/2021-07-13-hierfit/" rel="next">HieRFIT: A hierarchical cell type classification tool for projections from complex single-cell atlas datasets </a>
  </div>
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/publication/2021-03-11-chips-snakemake/" rel="prev">CHIPS: A Snakemake pipeline for quality control and reproducible processing of chromatin profiling data</a>
  </div>
  
</div>

</div>


<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 Ming Tang &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    
    <script id="dsq-count-scr" src="//divingintogeneticsandgenomics.disqus.com/count.js" async></script>
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/go.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/bash.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/shell.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

