---
title: The end of 2018
author: Ming Tang
date: '2018-12-28'
slug: the-end-of-2018
categories:
  - R
  - review
tags:
  - blogdown
header:
  caption: ''
  image: ''
---

It is almost the end of 2018. It is a good time to review what I have achieved during the year
and look forward to a brand new 2019. I wrote a similar post for 2017 [here](http://crazyhottommy.blogspot.com/2017/12/).

### Some highlights of the year 2018:

* My son Noah Tang was born in April. He is so lovely and we love him so much. Can't believe he is 
almost 9 months old.
![](/img/noah.jpg)
* Our epigenomic project was selected by the [Data Science Road-Trip program](https://bigdatau.ini.usc.edu/roadtrip) by USC. I spent 2 weeks in PNNL and worked closely with [Lisa Bramer](https://www.pnnl.gov/science/staff/staff_info.asp?staff_num=8785) and developed a pipeline to do feature selection using machine learning from a lot of chromHMM data sets. You can find the github repo [here](https://github.com/crazyhottommy/pyflow-chromForest/tree/vsurf_merge). I kept a note for everyday what I did as well at [here](https://github.com/crazyhottommy/Epigenome_RoadTrip). I will think about writing it up.

* I finally migrated my [previous blog](http://crazyhottommy.blogspot.com/) to blogdown which you are reading now :) Oh my, I love it. It makes blogging so much fun.

* I taught the ChIP-seq lesson for [2018 ANGUS Next-Gen Sequence Analysis Workshop](https://divingintogeneticsandgenomics.rbind.io/talk/2018-dibsi-course/) held in UC Davis from 7/1/2018 to 7/14/2018, and TAed for the rest of the sessions. It was a great teaching experience for me. I got to know many people and built connections. Most importantly, I enjoyed the teaching very much! Thanks [Titus Brown](https://biology.ucdavis.edu/people/c-titus-brown) for the invitation. I highly recommend you to attend this workshop if you want to start learning sequencing data analysis. The environment is so welcoming and Titus is so hilarious:) I was in the workshop to learn in 2014 and now I am back to teach! 

* I went back to University of Florida where I did my PhD to give a [talk](https://divingintogeneticsandgenomics.rbind.io/talk/2018-uf-talk/). It was very nice to be back home and catch up with my supervisor, other professors and some church friends! 

* Several co-author [papers/see the publications section](https://divingintogeneticsandgenomics.rbind.io/) are out in 2018. My first video shot for JOVE can be found at https://www.jove.com/video/56972/an-integrated-platform-for-genome-wide-mapping-chromatin-states-using. I was nervous but It was fun.

* In addition, two papers are out in Biorxiv at the end of 2018. I am co-first author in one of them. Both papers describe how epigenetic regulator KMT2D mediate tumor progression in melanoma and lung cancer, respectively. We are in the process submitting those papers to journals, but you can read more details at https://www.biorxiv.org/content/early/2018/12/28/507202 and https://www.biorxiv.org/content/early/2018/12/27/507327.  
The work was done in [Kunal Rai's lab](http://railab.org/people.html) where I had a chance to play with large amount of ChIP-seq data sets. My [Snakemake pipeline](https://divingintogeneticsandgenomics.rbind.io/project/snakemake-pipelines/) is being used in the lab by others and has processed thousands of ChIP-seq data sets.

* My three [most stared github repos](https://github.com/crazyhottommy) were cited in a paper [GitHub Statistics as a Measure of the Impact of Open-Source Bioinformatics Software](https://www.frontiersin.org/articles/10.3389/fbioe.2018.00198/full) by [Mikhail G. Dozmorov](https://medschool.vcu.edu/expertise/detail.html?id=mdozmorov). The table summarizing the popular github repos can be found at [here](https://github.com/mdozmorov/bioinformatics-impact/blob/master/tables/table_1.md)

* I joined Harvard FAS Informatics as a bioinformatics scientist in October and started working on single-cell RNAseq with [Dulac lab](https://www.dulaclab.com/) and will have a chance to play with other single molecule transcriptome data generated from [Xiaowei Zhuang's lab](http://zhuang.harvard.edu/). I am really excited about learning more!

* I have my green card approved. This is important so I can work in the US without worrying about my visa status. Thanks to my previous postdoc adviser [Roel Verhaak](https://www.jax.org/research-and-faculty/faculty/roel-verhaak), [Luca Pinello](http://pinellolab.org/), [Istvan Albert](https://www.ialbert.me/),  [Praveen Sethupathy](https://blogs.cornell.edu/sethupathylab/) and [Liang Cheng](https://medicine.iu.edu/faculty/14584/cheng-liang/) for writing recommendation letters. I will need to visit Luca's lab to say thanks personally, I have not had a chance to meet him after I moved to Harvard.

* I finally got a chance to write an R package. It is now in github private repo and works. I implemented some functions for visualizing single cell data and evaluating cluster stability. I will make it public once I clean up a bit. I was so satisfied to have it installed by `devtools::install_github()` and all functions and help pages are readily available as a package. I mean I am gradually transiting myself from an R user to R programmer. I know I still have a lot to learn, but this is exciting! By the way, I highly recommend the [`usethis`](https://github.com/r-lib/usethis) package for writing R packages and read the [`R pacakges book`](http://r-pkgs.had.co.nz/) by Hadley Wickham. read [this](https://blog.methodsconsultants.com/posts/developing-r-packages-using-gitlab-ci-part-i/) and [this](https://www.hvitfeldt.me/blog/usethis-workflow-for-package-development/) blog post to get started on how to write a minimal functional R package.

### What to expect in 2019

* I will have a lot of opportunities to teach workshops on R, Unix and single-cell RNAseq at my current position.
* A lot to learn on neuroscience. I will audit classes taught by Catherine Dulac.
* I will guest lecture a few lessons for [STAT 115: Introduction to Computational Biology and Bioinformatics](https://canvas.harvard.edu/courses/39391) taught by [Sheirly Liu](http://liulab.dfci.harvard.edu/).
* I would love to learn more on deep learning and apply it to single cell data analysis. Currently, I am taking classes from Coursera.
* Attend several conferences. Would love to catch up with the twitter-verse in person.
* A few more papers to write. I have at least 2 first author papers to finish. It's hanging there forever.
* Of course, spend time with the family.


