---
title: Mixing mouse and human 10x single cell RNAseq data
author: Ming Tang
date: '2019-12-11'
slug: mixing-mouse-and-human-10x-single-cell-rnaseq-data
categories:
  - scRNAseq
  - bioinformatics
tags:
  - scRNAseq
header:
  caption: ''
  image: ''
---



<p>In a typical “barnyard” experiment in which cells from different species are mixed before loading to the 10x controller, the identification of the species of origin after mapping/counting with the hybrid reference is a problem. People tend to use the ratio of reads mapped to each reference genome to determine which species a cell is from.</p>
<p>In this paper <a href="https://www.biorxiv.org/content/10.1101/630087v1.full" class="uri">https://www.biorxiv.org/content/10.1101/630087v1.full</a></p>
<blockquote>
<p>To deconvolute species, detect doublets and low quality cells, the mixed-species mapped data was used. Cells for which &gt;70% of the reads mapped to only one species were assigned to the corresponding species. The remaining cells (those for which &lt;70% of the reads mapped to only one species) were removed from the downstream analysis.</p>
</blockquote>
<p>kallisto bustool <a href="https://bustools.github.io/BUS_notebooks_R/10xv2.html">https://bustools.github.io/BUS_notebooks_R/10xv2.html</a> uses 90% as a cutoff.</p>
<p>However, how to choose this cutoff is subjective. In this blog post, I will mix a mouse and a human 10x single cell RNAseq dataset in silicon and then map to the hybrid transriptome. We have the ground truth of which cell comes from which species so we can investigate the mapping rate.</p>
<div id="download-data" class="section level3">
<h3>Download data</h3>
<p>Download 1k pbmc data and 1k mouse brain data from 10x website.</p>
<pre class="bash"><code>cd /n/holylfs/LABS/informatics/mtang/projects/dj/10x_mouse_human_mixing
wget http://cf.10xgenomics.com/samples/cell-exp/3.0.0/neuron_1k_v3/neuron_1k_v3_fastqs.tar

wget http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_fastqs.tar

## after tar xvf
ls neuron_1k_v3_fastqs/
neuron_1k_v3_S1_L001_I1_001.fastq.gz  neuron_1k_v3_S1_L001_R2_001.fastq.gz  neuron_1k_v3_S1_L002_R1_001.fastq.gz
neuron_1k_v3_S1_L001_R1_001.fastq.gz  neuron_1k_v3_S1_L002_I1_001.fastq.gz  neuron_1k_v3_S1_L002_R2_001.fastq.gz

ls pbmc_1k_v3_fastqs/
pbmc_1k_v3_S1_L001_I1_001.fastq.gz  pbmc_1k_v3_S1_L001_R2_001.fastq.gz  pbmc_1k_v3_S1_L002_R1_001.fastq.gz
pbmc_1k_v3_S1_L001_R1_001.fastq.gz  pbmc_1k_v3_S1_L002_I1_001.fastq.gz  pbmc_1k_v3_S1_L002_R2_001.fastq.gz</code></pre>
</div>
<div id="add-species-barcode-to-the-r1-fastq" class="section level3">
<h3>add species barcode to the R1 fastq</h3>
<p>Different experiment could have barcode collisions, let’s add additional barcode in front of the original cell barcode.</p>
<p>save the below as a <code>add_species_barcode.sh</code>:</p>
<p>This script adds a 8 base pair barcode in front of the R1 reads (16bp cell barcode + 12 bp umi for 10x version3) and some dummy high quality score (I) to the quality line.</p>
<pre class="bash"><code>#! /bin/bash
set -euo pipefail

zcat $1 |  awk -v barcode=&quot;$2&quot; &#39;NR%4 == 2 {$0=barcode$0} NR%4==0 {$0=&quot;IIIIIIII&quot;$0} {print}&#39; | gzip  &gt; $3</code></pre>
<p>I am adding <code>AAAAAAAA</code> and <code>TTTTTTTT</code> to mouse and human data, respectively.</p>
<pre class="bash"><code>chmod u+x add_species_barcode.sh

./add_species_barcode.sh neuron_1k_v3_fastqs/neuron_1k_v3_S1_L001_R1_001.fastq.gz AAAAAAAA neuron_1k_v3_fastqs/neuron_1k_v3_S1_L001_R1_001_modified.fastq.gz


./add_species_barcode.sh neuron_1k_v3_fastqs/neuron_1k_v3_S1_L002_R1_001.fastq.gz AAAAAAAA neuron_1k_v3_fastqs/neuron_1k_v3_S1_L002_R1_001_modified.fastq.gz


./add_species_barcode.sh pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R1_001.fastq.gz TTTTTTTT pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R1_001_modified.fastq.gz


./add_species_barcode.sh pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R1_001.fastq.gz TTTTTTTT pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R1_001_modified.fastq.gz
</code></pre>
</div>
<div id="build-a-hybrid-index-for-kallisto." class="section level3">
<h3>build a hybrid index for kallisto.</h3>
<p>kallisto uses cDNA for indexing.</p>
<p>Note <a href="https://www.kallistobus.tools/kb_transcriptome_index.html"><code>kb-python</code></a> uses genomics DNA and gtf file for making reference (it will extract the cDNA from genomic DNA based on gtf file). I tested <code>kb-python</code> for a single species experiment and it worked well and saves you doing all the steps. However, <code>kb ref</code> requires the fasta and gtf files to be merged for creating the hybrid reference. I may try it next time.</p>
<pre class="bash"><code>## download the cDNA for mouse and human
wget ftp://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz
wget ftp://ftp.ensembl.org/pub/release-96/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz

## download the gtf files
wget ftp://ftp.ensembl.org/pub/release-96/gtf/mus_musculus/Mus_musculus.GRCm38.96.gtf.gz
wget ftp://ftp.ensembl.org/pub/release-96/gtf/homo_sapiens/Homo_sapiens.GRCh38.96.gtf.gz

kallisto index -i GRCh38_GRCm38/GRCh38_GRCm38_96.idx Homo_sapiens.GRCh38.cdna.all.fa.gz Mus_musculus.GRCm38.cdna.all.fa.gz</code></pre>
<div id="kallisto-count-at-transcript-level" class="section level4">
<h4>kallisto count at transcript level</h4>
<p>Note 10x V2 R1 is 16bp cell barcode + 10 bp umi, V3 R1 is 16 bp cell barcode + 12 bp umi.</p>
<p>See my previous post <a href="https://divingintogeneticsandgenomics.rbind.io/post/understand-10x-scrnaseq-and-scatac-fastqs/" class="uri">https://divingintogeneticsandgenomics.rbind.io/post/understand-10x-scrnaseq-and-scatac-fastqs/</a></p>
<pre class="bash"><code>
kallisto bus -i /n/holylfs/INTERNAL_REPOS/INFORMATICS/reference_genome_by_tommy/kallisto_bus_ref/GRCh38_GRCm38_96.idx -o mouse_human_kallisto_out -x 0,0,24:0,24,36:1,0,0 -t8 \
neuron_1k_v3_fastqs/neuron_1k_v3_S1_L001_R1_001_modified.fastq.gz \
neuron_1k_v3_fastqs/neuron_1k_v3_S1_L001_R2_001.fastq.gz \
neuron_1k_v3_fastqs/neuron_1k_v3_S1_L002_R1_001_modified.fastq.gz \
neuron_1k_v3_fastqs/neuron_1k_v3_S1_L002_R2_001.fastq.gz \
pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R1_001_modified.fastq.gz \
pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R2_001.fastq.gz \
pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R1_001_modified.fastq.gz \
pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R2_001.fastq.gz

[index] k-mer length: 31
[index] number of targets: 307,242
[index] number of k-mers: 208,670,671
[index] number of equivalence classes: 1,276,238
[quant] will process sample 1: neuron_1k_v3_fastqs/neuron_1k_v3_S1_L001_R1_001_modified.fastq.gz
                               neuron_1k_v3_fastqs/neuron_1k_v3_S1_L001_R2_001.fastq.gz
[quant] will process sample 2: neuron_1k_v3_fastqs/neuron_1k_v3_S1_L002_R1_001_modified.fastq.gz
                               neuron_1k_v3_fastqs/neuron_1k_v3_S1_L002_R2_001.fastq.gz
[quant] will process sample 3: pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R1_001_modified.fastq.gz
                               pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R2_001.fastq.gz
[quant] will process sample 4: pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R1_001_modified.fastq.gz
                               pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R2_001.fastq.gz
[quant] finding pseudoalignments for the reads ... done
[quant] processed 159,504,118 reads, 97,750,679 reads pseudoaligned</code></pre>
</div>
</div>
<div id="bustools-count-at-gene-level" class="section level3">
<h3>bustools count at gene level</h3>
<p>we need a transcript to gene mapping tsv file making from gtf file. I could not find the <code>t2g.py</code> script mentioned in the <a href="https://www.kallistobus.tools/documentation">tutorial</a>.</p>
<p>make a transcript to gene mapping file using unix command line. also read my previous blog post: <a href="https://divingintogeneticsandgenomics.rbind.io/post/how-to-make-a-transcript-to-gene-mapping-file/" class="uri">https://divingintogeneticsandgenomics.rbind.io/post/how-to-make-a-transcript-to-gene-mapping-file/</a></p>
<pre class="bash"><code>zcat Homo_sapiens.GRCh38.96.gtf.gz | grep -v &quot;#&quot; | awk &#39;$3==&quot;transcript&quot;&#39; | cut -f9 | tr -s &quot;;&quot; &quot; &quot; | awk -v OFS=&quot;\t&quot; &#39;{print$6&quot;\t&quot;$2&quot;\t&quot;$10}&#39; | sort | uniq |  sed &#39;s/\&quot;//g&#39; &gt; Homo_sapiens.GRCh38.96.tsv


zcat Mus_musculus.GRCm38.96.gtf.gz | grep -v &quot;#&quot; | awk &#39;$3==&quot;transcript&quot;&#39; | cut -f9 | tr -s &quot;;&quot; &quot; &quot; | awk -v OFS=&quot;\t&quot; &#39;{print$6&quot;\t&quot;$2&quot;\t&quot;$10}&#39; | sort | uniq |  sed &#39;s/\&quot;//g&#39; &gt; Mus_musculus.GRCm38.96.tsv

## merge the tsv 
cat Homo_sapiens.GRCh38.96.tsv Mus_musculus.GRCm38.96.tsv &gt; GRCh38_GRCm38.96.tsv</code></pre>
</div>
<div id="correct-to-the-whitelist-and-bustools-count-at-gene-level" class="section level3">
<h3>correct to the whitelist and bustools count at gene level</h3>
<pre class="bash"><code>wget https://github.com/BUStools/getting_started/releases/download/species_mixing/10xv3_whitelist.txt

wc -l 10xv3_whitelist.txt
6794880 10xv3_whitelist.txt</code></pre>
<p>There are valid 68k cell barcodes from 10x. we added the species barcode in front of them and use bustool to correct for sequencing errors.</p>
<pre class="bash"><code># add the same species barcode to it.
cat 10xv3_whitelist.txt | awk &#39;{print &quot;AAAAAAAA&quot;$0}&#39; &gt; whitelist1.txt
cat 10xv3_whitelist.txt | awk &#39;{print &quot;TTTTTTTT&quot;$0}&#39; &gt; whitelist2.txt
cat whitelist1.txt whitelist2.txt &gt; whitelist.txt

mkdir tmp genecount

bustools correct -w whitelist.txt -p mouse_human_kallisto_out/output.bus | \
bustools sort -T tmp/ -t 4 -p - | \
bustools count -o genecount/genes \
-g /n/holylfs/INTERNAL_REPOS/INFORMATICS/reference_genome_by_tommy/kallisto_bus_ref/GRCh38_GRCm38.96.tsv \
-e mouse_human_kallisto_out/matrix.ec -t mouse_human_kallisto_out/transcripts.txt --genecounts -

Found 13589760 barcodes in the whitelist

Number of hamming dist 1 barcodes = 461228268
Processed 97750679 bus records
In whitelist = 94311259
Corrected = 329059
Uncorrected = 3110361
Read in 94640318 BUS records</code></pre>
<p><strong>Warning</strong> it is not working!! 0 genes were mapped when I checked the <code>genes.genes.txt</code>.</p>
<p>The devil is that <code>kallisto</code> infers the transcript id from the <code>cDNA</code> fasta file which contains the <code>.2</code> version number, but in the gtf file the version number is in the <code>transcript_version 2</code> entry.</p>
<p>The easiest way is to remove the version number in the <code>transcript.txt</code> file from output.</p>
<pre class="bash"><code># make a backup
cp transcripts.txt transcripts.bc.txt

# check how the version number look like for all genes
cat transcripts.txt | cut -d. -f2 | sort | uniq
1
10
11
12
13
14
15
16
17
2
3
4
5
6
7
8
9

cat transcripts.txt | sed -r &#39;s/\.[0-9]?//&#39; &gt; transcript2.txt

## rerun bustool but feeding the transcript2.txt
rm -rf genecount/*

bustools correct -w whitelist.txt -p mouse_human_kallisto_out/output.bus | \
bustools sort -T tmp/ -t 4 -p - | \
bustools count -o genecount/genes \
-g /n/holylfs/INTERNAL_REPOS/INFORMATICS/reference_genome_by_tommy/kallisto_bus_ref/GRCh38_GRCm38.96.tsv \
-e mouse_human_kallisto_out/matrix.ec -t mouse_human_kallisto_out/transcript2.txt --genecounts -</code></pre>
</div>
<div id="alternative-ways" class="section level3">
<h3>Alternative ways</h3>
<ol style="list-style-type: decimal">
<li>The R package <a href="https://github.com/BUStools/BUSpaRse/blob/master/R/tr2g.R">BUSpaRse</a> has a function to take care of that to make a transcript to gene mapping file from cDNA fasta. <a href="https://bustools.github.io/BUS_notebooks_R/10xv2.html" class="uri">https://bustools.github.io/BUS_notebooks_R/10xv2.html</a></li>
</ol>
<pre class="r"><code>library(BUSpaRse)
tr2g &lt;- transcript2gene(fasta_file = c(&quot;./data/hs_cdna.fa.gz&quot;, &quot;./data/mm_cdna.fa.gz&quot;),
                        kallisto_out_path = &quot;./output/out_hgmm1k&quot;)</code></pre>
<ol start="2" style="list-style-type: decimal">
<li><code>kb-python</code> command <code>kb ref</code> takes <strong>genomic DNA fasta</strong> and gtf files and makes the index and a transcript to gene mapping file on the fly.</li>
</ol>
</div>
<div id="load-in-to-r" class="section level3">
<h3>load in to R</h3>
<p>following <a href="https://bustools.github.io/BUS_notebooks_R/10xv2.html" class="uri">https://bustools.github.io/BUS_notebooks_R/10xv2.html</a></p>
<pre class="r"><code>library(BUSpaRse)
library(tidyverse)
library(DropletUtils)
library(Matrix)
library(Seurat)

res_mat &lt;- read_count_output(&quot;~/Downloads/genecount&quot;,name = &quot;genes&quot;, tcc = FALSE)

dim(res_mat)</code></pre>
<pre><code>## [1]  71600 731356</code></pre>
<p>remove some of the empty droplets</p>
<pre class="r"><code>tot_counts &lt;- Matrix::colSums(res_mat)
summary(tot_counts)</code></pre>
<pre><code>##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
##      0.00      1.00      1.00     34.89      5.00 142612.00</code></pre>
<pre class="r"><code># Compute barcode rank from Dropletutils
bc_rank &lt;- barcodeRanks(res_mat)

qplot(bc_rank$total, bc_rank$rank, geom = &quot;line&quot;) +
  geom_vline(xintercept = bc_rank$knee, color = &quot;blue&quot;, linetype = 2) +
  geom_vline(xintercept = bc_rank$inflection, color = &quot;green&quot;, linetype = 2) +
  annotate(&quot;text&quot;, y = 1000, x = 1.5 * c(bc_rank$knee, bc_rank$inflection),
           label = c(&quot;knee&quot;, &quot;inflection&quot;), color = c(&quot;blue&quot;, &quot;green&quot;)) +
  scale_x_log10() +
  scale_y_log10() +
  labs(y = &quot;Barcode rank&quot;, x = &quot;Total UMI count&quot;)</code></pre>
<p><img src="/post/2019-12-11-mixing-mouse-and-human-10x-single-cell-rnaseq-data_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
</div>
<div id="filter-the-cells" class="section level3">
<h3>filter the cells</h3>
<pre class="r"><code># Filter the matrix
res_mat &lt;- res_mat[, tot_counts &gt; bc_rank$inflection]
dim(res_mat)</code></pre>
<pre><code>## [1] 71600  2375</code></pre>
<p>Now, we get around 2000 cells.</p>
<pre class="r"><code>gene_species &lt;- ifelse(str_detect(rownames(res_mat), &quot;^ENSMUSG&quot;), &quot;mouse&quot;, &quot;human&quot;)
mouse_inds &lt;- gene_species == &quot;mouse&quot;
human_inds &lt;- gene_species == &quot;human&quot;
# mark cells as mouse or human
cell_species &lt;- tibble(n_mouse_umi = Matrix::colSums(res_mat[mouse_inds,]),
                       n_human_umi = Matrix::colSums(res_mat[human_inds,]),
                       tot_umi = Matrix::colSums(res_mat),
                       prop_mouse = n_mouse_umi / tot_umi,
                       prop_human = n_human_umi / tot_umi)</code></pre>
</div>
<div id="ground-truth" class="section level3">
<h3>ground truth</h3>
<pre class="r"><code>str_detect(colnames(res_mat), &quot;^AAAAAAAA&quot;) %&gt;% table()</code></pre>
<pre><code>## .
## FALSE  TRUE 
##  1169  1206</code></pre>
<pre class="r"><code>str_detect(colnames(res_mat), &quot;^TTTTTTTT&quot;) %&gt;% table()</code></pre>
<pre><code>## .
## FALSE  TRUE 
##  1206  1169</code></pre>
<p>We have 1206 mouse cells and 1169 human cells</p>
<pre class="r"><code>cell_species&lt;- cell_species %&gt;% 
  mutate(ground_truth = case_when(
    str_detect(colnames(res_mat), &quot;^AAAAAAAA&quot;) ~ &quot;mouse&quot;,
    str_detect(colnames(res_mat), &quot;^TTTTTTTT&quot;) ~ &quot;human&quot;
  )) 

p1&lt;- ggplot(cell_species, aes(x = ground_truth, y = prop_mouse)) +
  geom_boxplot(aes(color = ground_truth))

p2&lt;- ggplot(cell_species, aes(x = ground_truth, y = prop_human)) +
  geom_boxplot(aes(color = ground_truth))

p&lt;- cowplot::plot_grid(
  p1 + theme(legend.position=&quot;none&quot;),
  p2 + theme(legend.position=&quot;none&quot;),
  align = &#39;vh&#39;)

## add back the legend
legend &lt;- cowplot::get_legend(
  # create some space to the left of the legend
  p1 + theme(legend.box.margin = margin(0, 0, 0, 12))
)

# add the legend to the row we made earlier. Give it one-third of 
# the width of one plot (via rel_widths).
cowplot::plot_grid(p, legend, rel_widths = c(2, .4))</code></pre>
<p><img src="/post/2019-12-11-mixing-mouse-and-human-10x-single-cell-rnaseq-data_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<p>minimal and maximal proportion for mapping rate.</p>
<pre class="r"><code># min 95%
cell_species$prop_human[cell_species$ground_truth == &quot;human&quot;] %&gt;%
  range()</code></pre>
<pre><code>## [1] 0.9505135 0.9994325</code></pre>
<pre class="r"><code># max 5%
cell_species$prop_mouse[cell_species$ground_truth == &quot;human&quot;] %&gt;%
  range()</code></pre>
<pre><code>## [1] 0.0005675369 0.0494864613</code></pre>
<pre class="r"><code># min 97%
cell_species$prop_mouse[cell_species$ground_truth == &quot;mouse&quot;] %&gt;%
  range()</code></pre>
<pre><code>## [1] 0.9743096 1.0000000</code></pre>
<pre class="r"><code># max 2%
cell_species$prop_human[cell_species$ground_truth == &quot;mouse&quot;] %&gt;%
  range()</code></pre>
<pre><code>## [1] 0.00000000 0.02569043</code></pre>
<p>assign species of origin by the proportion</p>
<pre class="r"><code>cell_species&lt;- cell_species %&gt;% 
  mutate(ground_truth = case_when(
    str_detect(colnames(res_mat), &quot;^AAAAAAAA&quot;) ~ &quot;mouse&quot;,
    str_detect(colnames(res_mat), &quot;^TTTTTTTT&quot;) ~ &quot;human&quot;
  )) %&gt;%
  mutate(species = case_when(
    prop_mouse &gt; 0.9 ~ &quot;mouse&quot;,
    prop_human &gt; 0.9 ~ &quot;human&quot;,
    TRUE ~ &quot;mixed&quot;
  ))

table(cell_species$ground_truth, cell_species$species)</code></pre>
<pre><code>##        
##         human mouse
##   human  1169     0
##   mouse     0  1206</code></pre>
<p>This makes a 100% match as expected.</p>
</div>
<div id="seurat-for-dimension-reduction-and-visualization" class="section level3">
<h3>Seurat for dimension reduction and visualization</h3>
<pre class="r"><code>seu &lt;- CreateSeuratObject(res_mat, min.cells = 3) %&gt;% 
  NormalizeData(verbose = FALSE) %&gt;% 
  ScaleData(verbose = FALSE) %&gt;% 
  FindVariableFeatures(verbose = FALSE)

seu &lt;- AddMetaData(seu, metadata = cell_species$ground_truth, col.name = &quot;species&quot;)
seu &lt;- RunPCA(seu, verbose = FALSE, npcs = 30)
seu &lt;- RunTSNE(seu, dims = 1:20, check_duplicates = FALSE)
DimPlot(seu, reduction = &quot;pca&quot;, pt.size = 0.5, group.by = &quot;species&quot;)</code></pre>
<p><img src="/post/2019-12-11-mixing-mouse-and-human-10x-single-cell-rnaseq-data_files/figure-html/unnamed-chunk-19-1.png" width="672" /></p>
<pre class="r"><code>DimPlot(seu, reduction = &quot;tsne&quot;, pt.size = 0.5, group.by = &quot;species&quot;)</code></pre>
<p><img src="/post/2019-12-11-mixing-mouse-and-human-10x-single-cell-rnaseq-data_files/figure-html/unnamed-chunk-19-2.png" width="672" /></p>
</div>
<div id="conclusion" class="section level3">
<h3>conclusion</h3>
<ul>
<li><p>Using ratio of mapped reads for each cell to identify the cell of origin works pretty well for mouse and human mixtures. what if we use a more close species to human say chimpanzee or monkeys? Also, in a real experiment, one may have doublets from different species.</p></li>
<li><p>The mapping ratio is so drasticially different so a much smaller cutoff can still seperate the sepecies very well. This could be due to the cell type difference. It will be interesting to do the same experiment but with human pbmc and mouse pbmc cells.</p></li>
<li><p>I used the hybrid reference for mapping. How it will look like if I map the human cells with the mouse reference or map the mouse cells with the human reference?</p></li>
</ul>
</div>
